<!DOCTYPE html>
<html>

  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width initial-scale=1" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <title>Stochastic Modeling in Biology</title>
    <meta name="description" content="A simple, whitespace, helvetica based portfolio theme.
">

    <link rel="stylesheet" href="/css/main.css">
    <link rel="canonical" href="/2018/04/15/stochsim.html">

    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.7.2/css/all.min.css">
	<link rel="stylesheet" href="/css/academicons.css"/>
    <link href="/css/lightbox.css" rel="stylesheet">

    <link rel="stylesheet" href="/css/zenburn.css"/>

    <!-- collect tags -->
    
      <!-- collect tags from blog posts -->
<!-- source: https://longqian.me/2017/02/09/github-jekyll-tag/ --->







    

	<link type="application/atom+xml" rel="alternate" href="/feed.xml" title="martin-holub.github.io" />
</head>

  
  <!-- Google tag (gtag.js) -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=G-1ZBQKKLMRB"></script>
  <script>
	window.dataLayer = window.dataLayer || [];
	function gtag(){dataLayer.push(arguments);}
	gtag('js', new Date());

	gtag('config', 'G-1ZBQKKLMRB');
  </script>

  <script type="text/javascript"
    src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
	</script>

  <body>

    <header class="site-header">

  <div class="wrapper">

    <nav class="site-nav">

      <div class="trigger">
        <!-- martin-holub.github.io instead of blog -->
        <a class="page-link" href="/">about</a>

        <!-- ( for page in site.pages )  was previously-->
        
          
          <a class="page-link" href="/blog">blog</a>
          
        
          
          <a class="page-link" href="/subscribe">subscribe</a>
          
        

      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <header class="post-header">
    <h1 class="post-title">Stochastic Modeling in Biology</h1>
    <p class="post-meta">April 15, 2018 — 10:42 • <a href="https://martinholub.github.io/2018/04/15/stochsim.html#disqus_thread">0 Comments</a></p>
    <span class="post-meta">[
      
        
          <a href="/tag/ethz"><code class="highligher-rouge"><nobr>ethz</nobr></code>&nbsp;</a>
      
        
          <a href="/tag/projects"><code class="highligher-rouge"><nobr>projects</nobr></code>&nbsp;</a>
      
        
          <a href="/tag/code"><code class="highligher-rouge"><nobr>code</nobr></code>&nbsp;</a>
      
    ]</span>
  </header>
  
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["\\(","\\)"] ],
      displayMath: [ ['$$','$$'] ],
      processEscapes: false,
      skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
    },
    messageStyle: "normal",
    "HTML-CSS": {
      preferredFont: null
    },
    TeX: {
      extensions: ["AMSmath.js", "AMSsymbols.js","AMScd.js", "autobold.js", "begingroup.js", "mhchem.js"]
    }
  });
</script>
<script
  type="text/javascript"
  charset="utf-8"
  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"
>
</script>


  <article class="post-content">
    <div class="img_row">
	<img class="col three" src="/img/stochsim_header2.png" alt="" title="header_img" />
</div>
<div class="col three caption">
<a href="http://www.elowitz.caltech.edu/" target="_blank">Source</a>
</div>

<h2 id="introduction">Introduction</h2>
<p><br />
Genetically engineered cells and organisms are being used to produce array of commonplace commercial products including <a href="http://www.madehow.com/Volume-7/Insulin.html">drugs</a> and <a href="https://www.bio.org/articles/current-uses-synthetic-biology">materials</a>. Biological engineering is being employed to enhance nutrition content of various foods and yield of crops. Recently, <a href="https://en.wikipedia.org/wiki/Cultured_meat">lab-grown meat</a> represents an interesting alternative to traditional production and <a href="https://www.nature.com/news/2010/100120/full/463288a.html">synthetic biology</a> is being <a href="https://www.ginkgobioworks.com/2018/04/11/creative-in-residence/">explored even by artists</a>.  These successes are, however, just baby steps in face of the richness and complexity of “products” nature is so adept in engineering. Like what you ask? Like YOU, for example.</p>

<p>In the meantime, development is ongoing at rapid pace. We are getting finer understanding of molecular-scale processes by which the life is implemented and by extension, we are becoming more adept in exercising control over them. Biology is becoming easier to analyze and design using established engineering approaches. At different system levels, from molecular to cellular to populational, mathematical modeling and abstraction in design are becoming possible. (<em>If this abstraction thing is too abstract for you, think about this as if in your Arduino project (like <a href="http://www.instructables.com/id/The-EyeWriter-20/">this one</a>) you had to build and debug the microcontroller first. Before you could do that, you would have to do the same for all the resistors, transistors and other elements of the board. Instead, and luckily, you just need to have the board shipped and plug it into your laptotp to boot it up. You don’t have to think of all the possibly very intricate details that make it work and can focus on building some application on top of it. This is the power of abstraction in design when put to work.</em>).</p>

<p>This is exciting, fun and hugely relevant because, take it from <del>Elon Musk</del> Steve Jobs:</p>

<blockquote>
  <p>[…] the biggest innovations of the 21st century will be at the intersection of biology and technology.</p>
</blockquote>

<p>Of course this topic is about as broad as an average statement of a politician so I will not try to cover it all here (<em>If you are interested in some slightly more scientific peek into the topic, a good place to start is <a href="https://www.youtube.com/watch?v=lNttxYdGHs4">this video</a> where Chris Voigt first gives general of the field and later explains his idea of programming genetic circuits the way we program computers</em>), rather I want to focus on the mathematical modeling part. I find it really nice how this helps me to think about the biological process in more rigorous and engineering way and believe you may benefit from this as well.</p>

<p>First, I will give you lightweight introduction to some theory behind and then show some code I wrote for simulating simple transcription and translation processes. It is all on <a href="https://github.com/martinholub/gillespie-notebook">github</a> so feel free to grab it afterwards to have some fun with it.</p>

<h2 id="biochemical-reaction-systems">Biochemical Reaction Systems</h2>
<p><br /></p>

<div class="img_row">
    <a href="/img/copy_numbers.png" target="_blank">
	<img class="col three" src="/img/copy_numbers.png" alt="Copy Numbers" title="Copy Numbers" /></a>
</div>
<div class="col three caption">
Illustrative example of copy numbers at different levels of biochemical hierarchy. Low copy numbers, i.e. these at genome to proteome levels, are affected by stochasticity and cannot be faithfully modeled using Ordinary Differential Equation (ODE) models.
</div>

<p>In what follows we are looking at distinct biochemical species present in a cell. Additionally, these species are present in low <a href="https://medical-dictionary.thefreedictionary.com/copy+number">copy numbers</a> that are significantly influenced by stochastic (random) events. You can think about this as having a big parking lot under your kitchen window (poor you) and not having a calendar. In the morning when you make your breakfast, you can very reliably say if it is a normal working day or a public holiday based on occupancy of the parking lot. Conversely, you friend, who has just two parking places within sight of his kitchen window cannot make a good guess of whether it is a public holiday or not based on if the two spots are empty. This could just as well be because somebody’s child got sick (which is as random event as it gets).</p>

<p>In biology, there are multiple sources of such randomness, including external environmental perturbations and inherent stochasticity stemming from statistical thermodynamics of molecular reactions. To account for this, we leave the familiar realm of concentrations and enter the world of single-molecule counts. In mathematical terms this means, that instead of writing differential equations for average concentration levels, we keep track of individual reaction events that have some probability of occurring on some time interval. Conceptually, the modeling framework is depicted in the following figure:</p>

<div class="img_row">
    <a href="/img/modelling_framework.png" target="_blank">
	<img class="col three" src="/img/modelling_framework.png" alt="Modeling Framework" title="Modeling Framework" /></a>
</div>
<div class="col three caption">
Conceptual modeling framework for biochemical reaction systems. <it>TX</it> and <it>TL</it> refer to transcription and translation respectively. (<a href="http://www.cds.caltech.edu/~murray/BFSwiki/index.php/Main_Page" target="_blank">Source</a>)
</div>

<h3 id="the-chemical-master-equation">The Chemical Master Equation</h3>
<p><br /></p>

<p>We want to resolve dynamics of a system at very fine scale both in space and time. Further, the system does not have equilibrium configuration. It turns out that we are interested in probability $P(\mathbf{q}, t)$ of a system to be in <em>microstate</em> $\mathbf{q}$ at time $t$. Microstate is somewhat elusive term, but in the case of <em>chemical reaction systems</em> this is just a vector of number of molecules of each species in the system.</p>

<p>Next, we assume that there are $M$ reactions $R_j$, $j = 1, …, M$ in total and that $\boldsymbol{\nu}_j$ is a change in state (i.e. change in number of molecules, e.g. $\pm 1$ for unimolecular reaction) associated with reaction $R_j$. If we look at sufficiently small time interval, we can write $a_j (\mathbf{q})dt$ for the probability that a reaction $j$ “fires” in interval $dt$, given that current microstate of the system is $\mathbf{q}$. The $a_j (\mathbf{q})$ is called <em>propensity function</em> and we have dropped the time dependence as we assume this is accounted for by the dependence on microstate (that itself changes in time).</p>

<p>The above definitions allow us to calculate the change of the probability distribution function over an interval $dt$:</p>

\[\begin{array}{lcl}
P(\mathbf{q}, t + dt ~|~ \mathbf{q}_0, t_0) &amp; = &amp;
P(\mathbf{q}, t ~|~ \mathbf{q}_0, t_0)\prod_{j=1}^{M}{(1-a_j(\mathbf{q})dt)} \\
&amp; + &amp; \prod_{j=1}^{M}{P(\mathbf{q} - \boldsymbol{\nu}_j, t ~|~\mathbf{q}_0, t_0)(a_j(\mathbf{q} -\boldsymbol{\nu}_j)dt)} \\
&amp; = &amp; P(\mathbf{q}, t ~|~ \mathbf{q}_0, t_0) \\
&amp; + &amp; \sum_{j=1}^{M}{\bigg(P(\mathbf{q} - \boldsymbol{\nu}_j, t ~|~\mathbf{q}_0, t_0)a_j(\mathbf{q} -\boldsymbol{\nu}_j) - P(\mathbf{q}, t ~|~ \mathbf{q}_0, t_0)a_j(\mathbf{q})\bigg)dt} \\
&amp; + &amp; \mathscr{O}(dt^{2})\\
\end{array}\]

<p>This may look scary but it really just states that the probability to get to state $\mathbf{q}$ is equal to probability that we are already in that state + the probability that we are one reaction away from that state and the matching reaction fires in the interval $dt$.
After dropping higher order terms in $dt$ and simplifying the notation we get:</p>

\[\frac{\partial P}{\partial t} (\mathbf{q}, t) = \sum_{j=1}^{M}{\bigg(P(\mathbf{q} - \boldsymbol{\nu}_j, t)a_j(\mathbf{q} -\boldsymbol{\nu}_j) - P(\mathbf{q}, t)a_j(\mathbf{q})\bigg)}.\]

<p>This is the <em>chemical master equation (CME)</em>, which is an example of <em>forward Kolmogorov equation</em> for a discrete state, continuous time random process (phew, quite some terminology here). The discrete state here is described by the vector $\mathbf{q}$ with numbers of molecules of species in the system. To be able to calculate anything with it we will need to specify the initial condition $P(\mathbf{q}_0, t_0)$. To help you put this into bigger picture, see the two following figures that show where the CME is positioned with respect to other (more approximate) approaches to model biological system dynamics.</p>

<div class="img_row">
    <a href="/img/modeling_approaches1.png" target="_blank">
	<img class="col three" src="/img/modeling_approaches1.png" alt="Modeling Approaches 1" title="Modeling Approaches 1" /></a>
</div>
<div class="col three caption">
Different methods for modeling biochemical reaction systems. (<a href="http://www.cds.caltech.edu/~murray/BFSwiki/index.php/Main_Page" target="_blank">Source</a>)
</div>

<div class="img_row">
    <a href="/img/modeling_approaches2.png" target="_blank">
	<img class="col three" src="/img/modeling_approaches2.png" alt="Modeling Approaches 2" title="Modeling Approaches 2" /></a>
</div>
<div class="col three caption">
Hierarchy of methods for modeling biochemical reactions. CME - Chemical Master Equation, SSA - Stochastic Simulation Algorithm, CLE - Chemical Langevin Equation, CFPE - Chemical Fokker-Planck Equation, RRE - Reaction Rate Equation. (<a href="https://www.annualreviews.org/doi/abs/10.1146/annurev.physchem.58.032806.104637" target="_blank">Source</a>)
</div>

<p>The main reason for why there is such a wide palette of approaches is because the chemical master equation does not have analytical solution and its numerical solution quickly becomes computationally expensive for systems with more than few distinct species. Its advantage is that building a model from set of reactions is very transparent and straightforward and this is why we will focus on precisely this approach, keeping in mind its limitations.</p>

<h2 id="the-stochastic-simulation-algorithm">The Stochastic Simulation Algorithm</h2>
<p><br />
We have already mentioned that <em>CME</em> doesn’t have analytical soltion. The straightforward approach how to obtain some representative description of the system is to draw samples from the probability distribution $P(\mathbf{q}, t ~|~ \mathbf{q}_0, t_0)$ governed by the <em>CME</em>. This is  <a href="https://www.quora.com/What-is-an-intuitive-explanation-of-Monte-Carlo-simulations"><em>Monte Carlo simulation</em></a>. We will be able to determine <a href="https://en.wikipedia.org/wiki/Moment_(mathematics)">moments of the distribution</a> (e.g. mean and variance) by averaging over many independent realizations of single trajectories (under identical initial conditions).
<!-- Can actually implement stochastic collocation here!! --></p>

<h4 id="state-update">State update</h4>
<p><br />
To implement this in code, we will follow approach as developed by <a href="https://en.wikipedia.org/wiki/Daniel_Gillespie">Daniel Gillespie</a>, i.e. the <em>stochastic simulation algorithm (SSA)</em>. Given an arbitrary state $\mathbf{q} = (n_1, n_2, …)$, for computation of the next state we need an index of next reaction that occurs (“fires”) and $\Delta t$ in which this happens. The probability that the next reaction to fire is the reaction $j$ is directly proportional to the corresponding propensity $a_j$ (intuitively, reactions with high reaction rates will occur more often). Therefore finding an index $j$ of such reaction is just matter of drawing a random integer from a weighted distribution on $[1, M]$ where the weights are given as:</p>

\[\frac{a_j}{\sum_{i}^{M} a_i} \qquad j = 1, ..., M.\]

<p>Note that this is equivalent to drawing a uniform random number on $[0,1]$, scaling it by the total propensity $a = \sum_{i}^{M}a_i$ and finding the first integer $j$ for which this scaled random number is less then $\sum_{i}^{j}a_i$.</p>

<h4 id="time-update">Time update</h4>
<p><br />
Finding a $\Delta t$ in which this reaction occurs is more involved. You can find the derivation at the <a href="#Time-step-size-derivation-for-SSA">end of this post</a>, but for now we will take for granted that the <em>cumulative distribution function (CDF)</em> for $\Delta t$ is:</p>

\[\int_{0}^{\Delta t} \mathrm{f}_T(\tau) d\tau = \int_{0}^{\Delta t} a \mathrm{e}^{-a\tau}d \tau = 1 - \mathrm{e}^{-a\Delta t}.\]

<p>This is nice because we draw samples from this distribution by drawing a random number $\xi$ on $[0, 1]$ and <a href="https://en.wikipedia.org/wiki/Inverse_transform_sampling">inverse-sampling</a> from the CDF. In this case, the CDF can be even inverted analytically to obtain:</p>

\[\Delta t = \frac{1}{a} \mathrm{ln}\bigg(\frac{1}{1 - \xi}\bigg).\]

<h4 id="python-code">Python code</h4>
<p><br />
If the previous two sections sounded complicated to you, have no worry, it is actually dead simple to implement them in code:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">draw_ssa</span><span class="p">(</span><span class="n">propensities</span><span class="p">,</span> <span class="n">total_prop</span><span class="p">):</span>
    <span class="s">"""Draw random samples according to prescriptions of SSA

    Parameters
    -------
    propensities: array, propensities of all reactions
    total_prop: float, total propensity

    Returns
    ---------
    idx: int, index into the state change matrix to determine updated
              state
    tau: float, time step size
    """</span>

    <span class="n">xi1</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">random</span><span class="p">.</span><span class="n">uniform</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">prop_rank</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">sort</span><span class="p">(</span><span class="n">np</span><span class="p">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">propensities</span><span class="p">)</span><span class="o">/</span><span class="n">total_prop</span><span class="p">)</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">int32</span><span class="p">(</span><span class="n">np</span><span class="p">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">prop_rank</span> <span class="o">&gt;</span> <span class="n">xi1</span><span class="p">))</span>

    <span class="n">xi2</span> <span class="o">=</span>  <span class="n">np</span><span class="p">.</span><span class="n">random</span><span class="p">.</span><span class="n">uniform</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">tau</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">float32</span><span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="n">total_prop</span> <span class="o">*</span> <span class="n">np</span><span class="p">.</span><span class="n">log</span><span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="n">xi2</span><span class="p">)))</span>
    <span class="k">return</span> <span class="n">idx</span><span class="p">,</span> <span class="n">tau</span>
</code></pre></div></div>

<h3 id="the-algorithm">The Algorithm</h3>
<p><br />
Now we are armed for the stochastic simulation algorithm that proceeds as follows:</p>

\[\begin{alignat}{0}
\textbf{initialize} \\
\qquad \textrm{choose}\ \mathbf{q}_0,\ \textrm{e.g.}\ (n_{mRNA,0}, n_{protein, 0})\\
\textbf{while}\ t\ &lt; \ t_{stop}:\\
\qquad \textrm{calculate}\ a_i(\mathbf{q}),\ \forall\ i = 1, ..., M \\
\qquad \textrm{draw}\ \xi_1 \in [0, 1]\ \textrm{and compute}\ \Delta t \\
\qquad \textrm{draw}\ \xi_2 \in [0, 1]\ \textrm{and find index}\ j \\
\qquad \textrm{update}\ \mathbf{q}\\
\end{alignat}\]

<p>This algorithm is also called <a href="https://www.sciencedirect.com/science/article/pii/0021999176900413?via%3Dihub">Gillespie’s direct method</a> as it was D.T. Gillespie who proved that this method samples from the probability distribution governed by the CME. The method is derived based on <a href="http://cmt.dur.ac.uk/sjc/thesis_mcg/node6.html">first principles</a>. This makes it an exact approach to stochastic simulation, consequently, time step $\tau$ is not mere finite approximation (as in ODE models).</p>

<h4 id="python-code-1">Python code</h4>
<p><br />
Implementation in code is again quite simple:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">ssa_routine</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">t_end</span><span class="p">,</span> <span class="n">q0</span><span class="p">,</span> <span class="n">nu</span><span class="p">,</span> <span class="n">psi</span><span class="p">):</span>
    <span class="s">""" Stochastic Simulation Algorithm

    Performs sampling from distribution governed by chemical master equation.

    Parameters
    --------------
    k: array, vector stochastic reaction constants
    t_end: int, number of steps for the simulation
    q0: array, initial number of molecules for each species
    nu: ndarray, stoichiometric matrix, with state change vectors in columns
    psi: ndarray, reactant molecularity matrix

    Returns
    --------------
    Q: array, all states q at times t
    time: array, vector of times t
    tot_props: array, total propensities at times t
    is_success: bool, flag indicating if simulation ran to completion
    """</span>

    <span class="c1">#Initialize arrays to store states, tick times
</span>    <span class="n">time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">t</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">time</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>

    <span class="n">Q</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">state</span> <span class="o">=</span> <span class="n">q0</span>
    <span class="n">Q</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="n">state</span><span class="p">)</span>

    <span class="k">while</span> <span class="n">t</span> <span class="o">&lt;</span> <span class="n">t_end</span><span class="p">:</span>
        <span class="c1"># Calculate propensities for current state
</span>        <span class="n">propensities</span><span class="p">,</span> <span class="n">total_prop</span> <span class="o">=</span> <span class="n">get_propensities</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">psi</span><span class="p">)</span>

        <span class="c1"># Draw random samples for update of state and time step
</span>        <span class="n">idx</span><span class="p">,</span> <span class="n">tau</span> <span class="o">=</span> <span class="n">draw_ssa</span><span class="p">(</span><span class="n">propensities</span><span class="p">,</span> <span class="n">total_prop</span><span class="p">)</span>

        <span class="c1">## Find random change in state - see theory
</span>        <span class="n">state</span> <span class="o">=</span> <span class="n">state</span> <span class="o">+</span> <span class="n">nu</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="p">:]</span>

        <span class="c1">## Find random time step size - see theory
</span>        <span class="n">t</span> <span class="o">=</span> <span class="n">t</span> <span class="o">+</span> <span class="n">tau</span>

    <span class="k">return</span> <span class="n">Q</span><span class="p">,</span> <span class="n">time</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">get_propensities</span><span class="p">(</span><span class="n">rates</span><span class="p">,</span> <span class="n">populations</span><span class="p">,</span> <span class="n">psi</span><span class="p">):</span>
    <span class="s">"""Return propensities given rates and populations

    Computes the value of propensity for all reactions. Generalizes to various systems

    Parameters
    ----------
    rates: array, vector of stochastic reaction constants
    populations: array, vector of number of molecules of each species

    Returns
    ---------------
    propensities: array, vector of propensiites for all reactions
    total_propensity: float, sum of all propensiites
    """</span>

    <span class="c1"># See model description
</span>    <span class="n">propensities</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">empty</span><span class="p">(</span><span class="n">psi</span><span class="p">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">float32</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">psi</span><span class="p">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">populations</span> <span class="o">**</span> <span class="n">psi</span><span class="p">[</span><span class="n">r</span><span class="p">,:]</span>
        <span class="n">propensities</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">.</span><span class="n">prod</span><span class="p">()</span> <span class="o">*</span> <span class="n">rates</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>

    <span class="n">total_propensity</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">float32</span><span class="p">(</span><span class="n">propensities</span><span class="p">.</span><span class="nb">sum</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">propensities</span><span class="p">,</span> <span class="n">total_propensity</span>
</code></pre></div></div>

<p>The apparent drawback of this method is its computational complexity. As the systems become of somewhat more practical size (larger number of different species), the time step becomes very small (reactions occur often). As on each time step, two random numbers have to be generated, this leads to significant computational cost. To circumvent this issue, approximations of the SSA  become useful (see Figure <em>Hierarchy of methods for modeling biochemical reactions</em>).</p>

<h2 id="easy-example-the-central-dogma">Easy example: The Central Dogma</h2>
<p><br />
To make this discussion more concrete, let’s look at the minimal example of <a href="https://www.dnalc.org/resources/3d/central-dogma.html"><em>central dogma of molecular biology</em></a>. This can be concisely represented as:</p>

\[\require{AMScd}
\begin{CD}
    DNA @&gt;\beta_m&gt;&gt; mRNA @&gt;{\beta_p \cdot n_m(t)}&gt;&gt; protein\\
    @. @VV{\gamma_m \cdot n_m(t)}V @VV{\gamma_p \cdot n_p(t)}V\\
    @. \varnothing @. \varnothing
\end{CD}\]

<p>Here the microstate is simply the vector $\mathbf{q} = (1, n_m, n_p)$, where $n_m$ and $n_p$ are the copy numbers of transcripts and proteins respectively that depend on time. The DNA is present in single copy throughout (as such, we do not need to include it in our calculations, but will do so for consistency). The propensities (or <em>transition probabilities</em>) are:</p>

\[a_1 = \beta_m \qquad a_2 = \gamma_m \cdot n_m \qquad a_3 = \beta_p \cdot n_m \qquad a_4 = \gamma_p \cdot n_p,\]

<p>where $\beta_i$ corresponds to <em>production</em> and $\gamma_i$ to <em>degradation</em>. We can see that propensity, in case of unimolecular reaction ($A \rightarrow B$) directly corresponds to the reaction rate constant (Things get only bit more complicated for bimolecular reactions as you will see later).</p>

<p>Each state change vector corresponds to one reaction in the system and together they are concatenated as columns of <em>stoichiometric matrix</em> $\mathbf{N}$:</p>

\[\mathbf{N}= (\boldsymbol{\nu}_1^T, \boldsymbol{\nu}_2^T, \boldsymbol{\nu}_3^T, \boldsymbol{\nu}_4^T) =
\begin{bmatrix}
0 &amp; 0 &amp; ~0 &amp; ~0 \\
1 &amp; 0 &amp; -1 &amp; ~0 \\
0 &amp; 1 &amp; ~0 &amp; -1
\end{bmatrix}\]

<p>The dynamics of the system are described by <em>Chemical master equation (CME)</em> for probability $P\big((n_m, n_p), t\big)$:</p>

\[\begin{array}{lcl}
\frac{\partial P}{\partial t}\big((n_m, n_p), t\big) &amp; = &amp; P\big((n_m-1, n_p\big), t)\beta_m - P\big((n_m, n_p), t\big)\beta_m \\
&amp; + &amp; P\big((n_m+1, n_p), t\big)\gamma_m (n_m+1) - P\big((n_m, n_p), t\big)\gamma_m n_m \\
&amp; + &amp; P\big((n_m, n_p-1), t\big)\beta_p (n_m) - P\big((n_m, n_p), t\big)\beta_p n_m \\
&amp; + &amp; P\big((n_m, n_p+1), t\big)\gamma_p (n_p+1) - P\big((n_m, n_p), t\big)\gamma_p n_p \\
\end{array}\]

<p>Phew! This has been quite some theory we have covered, let’s now take a look how we can translate all this into code to simulate a real biological system.</p>

<h3 id="simulation">Simulation</h3>
<p><br /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">%</span><span class="n">load_ext</span> <span class="n">autoreload</span>
<span class="o">%</span><span class="n">autoreload</span> <span class="mi">2</span>
<span class="kn">import</span> <span class="nn">ssa_routine</span> <span class="k">as</span> <span class="n">ssa</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="n">plt</span>

<span class="n">plt</span><span class="p">.</span><span class="n">style</span><span class="p">.</span><span class="n">use</span><span class="p">(</span><span class="s">'ggplot'</span><span class="p">)</span> <span class="c1"># use "ggplot" style for graphs
</span><span class="n">pltparams</span> <span class="o">=</span> <span class="p">{</span><span class="s">'legend.fontsize'</span><span class="p">:</span><span class="mi">14</span><span class="p">,</span><span class="s">'axes.labelsize'</span><span class="p">:</span><span class="mi">18</span><span class="p">,</span><span class="s">'axes.titlesize'</span><span class="p">:</span><span class="mi">18</span><span class="p">,</span>
             <span class="s">'xtick.labelsize'</span><span class="p">:</span><span class="mi">12</span><span class="p">,</span><span class="s">'ytick.labelsize'</span><span class="p">:</span><span class="mi">12</span><span class="p">,</span><span class="s">'figure.figsize'</span><span class="p">:(</span><span class="mf">7.5</span><span class="p">,</span><span class="mf">7.5</span><span class="p">),}</span>
<span class="n">plt</span><span class="p">.</span><span class="n">rcParams</span><span class="p">.</span><span class="n">update</span><span class="p">(</span><span class="n">pltparams</span><span class="p">)</span>
</code></pre></div></div>

<h4 id="model-parameters">Model Parameters</h4>
<p><br />
We set the initial conditions as following:</p>

<table>
  <thead>
    <tr>
      <th style="text-align: center">$DNA$</th>
      <th style="text-align: center">$mRNAs$</th>
      <th style="text-align: center">$Proteins$</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">1</td>
      <td style="text-align: center">0</td>
      <td style="text-align: center">0</td>
    </tr>
  </tbody>
</table>

<p>For the reaction rates we select:</p>

<table>
  <thead>
    <tr>
      <th style="text-align: center">$\beta_m$</th>
      <th style="text-align: center">$\beta_p$</th>
      <th style="text-align: center">$\gamma_m$</th>
      <th style="text-align: center">$\gamma_p$</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">10</td>
      <td style="text-align: center">10</td>
      <td style="text-align: center">1</td>
      <td style="text-align: center">0.4</td>
    </tr>
  </tbody>
</table>

<p>This means, for example, that the “birth” of a molecule of mRNA is 10-times more likely than its “death”.</p>

<p>Additionally we define matrix of reactant (species entering each reaction) molecularity $\mathbf{\psi}$:</p>

\[\mathbf{\psi}=
\begin{bmatrix}
1 &amp; 0 &amp; 0 &amp; 0 \\
0 &amp; 1 &amp; 1 &amp; 0 \\
0 &amp; 0 &amp; 0 &amp; 1
\end{bmatrix}.\]

<p>We will also simulate multiple independent evolutions (paths) such that we can obtain mean evolution.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mp</span> <span class="o">=</span> <span class="n">ssa</span><span class="p">.</span><span class="n">modelParameters</span><span class="p">(</span><span class="n">k</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">array</span><span class="p">([</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="p">.</span><span class="mi">4</span><span class="p">]),</span>
                         <span class="n">nu</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">array</span><span class="p">(([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">])),</span>
                         <span class="n">psi</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">array</span><span class="p">(([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])),</span>
                         <span class="n">q0</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]),</span>
                         <span class="n">t_end</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
                         <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s">"gene"</span><span class="p">,</span> <span class="s">"mRNA"</span><span class="p">,</span> <span class="s">"protein"</span><span class="p">],</span>
                         <span class="n">dist</span> <span class="o">=</span> <span class="p">[</span><span class="s">"none"</span><span class="p">,</span> <span class="s">"poiss"</span><span class="p">,</span> <span class="s">"gauss"</span><span class="p">]</span> <span class="p">)</span>
<span class="n">n_paths</span> <span class="o">=</span> <span class="mi">30</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">np</span><span class="p">.</span><span class="n">random</span><span class="p">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span>
<span class="n">ssa</span><span class="p">.</span><span class="n">main</span><span class="p">(</span><span class="n">mp</span><span class="p">,</span> <span class="n">n_paths</span><span class="p">,</span> <span class="n">do_load</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">do_save</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</code></pre></div></div>
<div class="language-html highlighter-rouge"><div class="highlight"><pre class="highlight"><code>2018-04-15 09:33:39,299:ssa_routine:INFO: Starting simulation with n_paths=30
2018-04-15 09:34:14,563:ssa_routine:INFO: Calculating statistics for all species.
2018-04-15 09:34:15,376:ssa_routine:INFO: Simulation finished.
2018-04-15 09:34:21,556:ssa_routine:INFO: Program execution finished.
</code></pre></div></div>

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: center;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: center;">
      <th></th>
      <th>mean</th>
      <th>var</th>
      <th>CV</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>gene</th>
      <td>1.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>mRNA</th>
      <td>10.431053</td>
      <td>10.262420</td>
      <td>0.307112</td>
    </tr>
    <tr>
      <th>protein</th>
      <td>252.870719</td>
      <td>2027.642923</td>
      <td>0.178073</td>
    </tr>
  </tbody>
</table>
</div>

<h3 id="results">Results</h3>
<p><br />
We obtain statistics for distribution of number of species. The mRNAs reach steady-state value of around 10 copies whereas the proteins level off at around 260 copies. The $CV$ is a <a href="https://en.wikipedia.org/wiki/Coefficient_of_variation">coefficient of variation</a>, normalized measure of dispersion of a distribution defined as a ratio of standard deviation to mean $CV = \frac{\sigma}{\mu}.$</p>

<p>We can look at the development of species numbers in time in the following plot in the left panel. On the right hand is then all-time histogram estimate of distribution of copy numbers together with empirical cumulative distribution function and KDE - <a href="https://www.quora.com/What-is-kernel-density-estimation">kernel density estimate</a>. We observe that the distribution of number of mRNAs well follows the theoretical <a href="https://en.wikipedia.org/wiki/Poisson_distribution">Poisson distribution</a> with rate parameter $\lambda = \beta_m$.</p>

<div class="img_row">
    <a href="/img/easy_example_evolution.png" target="_blank">
	<img class="col three" src="/img/easy_example_evolution.png" alt="Easy Example" title="Easy Example" /></a>
</div>
<div class="col three caption">
Temporal evolution of species' copy numbers. Red line indicates mean evolution for 30 independent realizations. Right panel displays marginal distribution for all times both as empirical estimate (histogram and KDE - Kernel Density Estimate) and theoretical predictions (Poisson for mRNAs and Gaussian for proteins). Green line in the right panel represents empirical cumulative distribution function.
</div>

<h3 id="note-on-speed">Note on speed</h3>
<p><br />
Even in this simple version, the algorithm takes quite some time to finish. I looked into it and improved the performance using <a href="https://numba.pydata.org/">numba</a>. You can find details on this in the <a href="#Optimizing-the-sampling-in-SSA">supplementary note</a>.</p>

<h2 id="advanced-example-noise-induced-switch">Advanced Example: Noise Induced Switch</h2>
<p><br />
OK, once we have the simulation and visualization functions ready and once we can draw samples reasonably fast, we can start working on a more complex model. We will look at set of reactions that will allow us to simulate effects of noise on gene expression levels. In particular we are interested in rapid switching between two significantly different levels of expression. This interesting phenomenon, referred as bistability, is often present in real biological circuits but cannot be captured by a deterministic ODE model. Whereas in previous easy example we may have as well used a deterministic ODE to obtain good estimate of time evolution, in this case we need a stochastic model to get representative picture of system dynamics.</p>

<h3 id="the-model">The model</h3>
<p><br />
\(G_1 \xrightarrow{\beta_1} G_1 + P_1 \qquad \qquad P_1 \xrightarrow{\gamma_1} \varnothing \tag{1a, 1b}\)</p>

\[G_2 \xrightarrow{\beta_2} G_2 + 2\,P_2 \qquad \qquad P_2 \xrightarrow{\gamma_1} \varnothing \tag{2a, 2b}\]

\[\require{mhchem}
\ce{G1 + P2 &lt;=&gt;[{\kappa_{c1}^{+}}][{\kappa_{c1}^{-}}] \xi_{1}^{I}} \tag{3}\]

\[\ce{G2 + P1 &lt;=&gt;[{\kappa_{c2}^{+}}][{\kappa_{c2}^{-}}] \xi_{2}^{I_1}} \tag{4}\]

\[\ce{\xi_{2}^{I_1} + P1 &lt;=&gt;[{\kappa_{c3}^{+}}][{\kappa_{c3}^{-}}] \xi_{2}^{I_2}} \tag{5}\]

<ul>
  <li>
    <p>Reactions $(1a)$ and $(2a)$ represent expression of proteins $P_1$ and $P_2$ originating from genes $G_1$ and $G_2$ ($G_2$ codes for <a href="https://en.wikipedia.org/wiki/Messenger_RNA#Monocistronic_versus_polycistronic_mRNA">bicistronic</a> mRNA). They reflect the fact that the genes are left unchanged and available for transcription.</p>
  </li>
  <li>
    <p>$(2a)$ and $(2b)$ correspond to protein degradation, where reaction rate is proportional to current number of molecules of given protein $n_i, ~ i={P_1, P_2}$.</p>
  </li>
  <li>
    <p>In equation $(3)$, the expression of protein $P_1$ is inhibited by $P_2$ which binds to promoter region of $G_1$ and acts as <em>inhibitor</em>, forming complex $\xi_1^I$.</p>
  </li>
  <li>
    <p>The expression of from gene $G_2$ is repressed by protein $P_1$ which together form an intermediate transcription-inhibiting complex $\xi_2^{I_1}$ (reaction $(4)$)</p>
  </li>
  <li>
    <p>Protein $P_1$ is inactivated by binding to complex $\xi_2^{I_1}$ and forming $\xi_2^{I_2}$ (reaction $(5)$). Note that reactions $(4)$ and $(5)$ work antagonistically.</p>
  </li>
</ul>

<p>We will model <a href="https://en.wikipedia.org/wiki/Cooperativity">cooperativity</a> of the enzyme binding sites with:</p>

\[\kappa_{c3}^{+} = \sigma \kappa_{c2}^{+} \qquad \sigma &gt;&gt; 1.\]

<p>This essentially says that the inhibitory binding to $P_1$ becomes easier after the first complexation reaction. This, at the first sight not very interesting feature of the system, is actually essential in our simulation. Repression with cooperativity is namely <a href="https://www.sciencedirect.com/science/article/pii/S0022519300910683">one of ways how to achieve switch-like behavior</a> in biological systems.</p>

<!---
This is a very nice property of the system that can be analytically proved from deterministic ODE model. This however requires careful mathematical analysis, so we will not do it here (if you are interested checkout [this website](https://mathbio.colorado.edu/index.php/MBW:How_to_make_a_Biological_Switch)).
--->

<p>The system in question has 7 distinct biochemical species: $[G_1, P_1, G_2, P_2, \xi_{1}^{I}, \xi_{2}^{I_1}, \xi_{2}^{I_2}]$ and comprises of in total 10 reactions (as we consider each reversible reaction as two).</p>

<h4 id="stochastic-reaction-constants">Stochastic reaction constants</h4>
<p><br />
As previously we will define an array of stochastic reaction rates $k$. We will have to be careful about units here. First, we note that for <em>unimolecular</em> reactions, the reaction rates are identical to the deterministic ones with unit ${min}^{-1}$. This is good because we are already familiar with them from previous example.</p>

<p>For <em>bimolecular</em> (two species) reactions, the stochastic reaction rate relates to the reaction rate of deterministic model by scaling with inverse volume $\Omega$.</p>

\[\hat{k}_i = \frac{k_i}{\Omega}.\]

<p>To obtain same units as in the case of <em>unimolecular</em> reaction, we include also scaling by Avogardo’s number $N_A [mol^{-1}]$, such that $\Omega = V\,
N_A$ and $k_i = [M \cdot min^{-1}]$. $N_A$ is the old familiar $6.022\cdot10^{23} mol^{-1}.$ For the volume, we make a <a href="http://book.bionumbers.org/how-big-is-an-e-coli-cell-and-what-is-its-mass/">ballpark estimate</a> for <em>E.coli</em> as a cylinder with diameter $1\mu m$ and length $2\mu m$ yielding $V \approx 1.5 \mu m^3 = 1.5\cdot 10^{-14} dm^3$.</p>

<p>For our model, we will select the following values of stochastic reaction constants:</p>

<table>
  <thead>
    <tr>
      <th style="text-align: center">$\beta_1$</th>
      <th style="text-align: center">$\beta_2$</th>
      <th style="text-align: center">$\gamma_1$</th>
      <th style="text-align: center">$\gamma_2$</th>
      <th style="text-align: center">$\kappa_{c1}^{+}$</th>
      <th style="text-align: center">$\kappa_{c1}^{-}$</th>
      <th style="text-align: center">$\kappa_{c2}^{+} $</th>
      <th style="text-align: center">$\kappa_{c2}^{-}$</th>
      <th style="text-align: center">$\kappa_{c3}^{+}$</th>
      <th style="text-align: center">$\kappa_{c3}^{-}$</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">$100$</td>
      <td style="text-align: center">$1000$</td>
      <td style="text-align: center">$10$</td>
      <td style="text-align: center">$4$</td>
      <td style="text-align: center">$\frac{2}{\Omega}$</td>
      <td style="text-align: center">$1$</td>
      <td style="text-align: center">$\frac{4}{\Omega}$</td>
      <td style="text-align: center">$1$</td>
      <td style="text-align: center">$\frac{4\sigma}{\Omega}$</td>
      <td style="text-align: center">$1$</td>
    </tr>
  </tbody>
</table>

<p>The choice here is less arbitrary then it may look, because only certain parameters of the system will lead to noise-induced switch-like behavior. Here we require $\kappa_{c1}^{-} = \kappa_{c2}^{-} = \kappa_{c3}^{-}$, $2 \kappa_{c1}^{+} = \kappa_{c2}^{+}$ and $\sigma = 10$.</p>

<p>This leads to following set of propensities:</p>

<table>
  <thead>
    <tr>
      <th style="text-align: center">$a_1$</th>
      <th style="text-align: center">$a_2$</th>
      <th style="text-align: center">$a_3$</th>
      <th style="text-align: center">$a_4$</th>
      <th style="text-align: center">$a_5$</th>
      <th style="text-align: center">$a_6$</th>
      <th style="text-align: center">$a_7$</th>
      <th style="text-align: center">$a_8$</th>
      <th style="text-align: center">$a_9$</th>
      <th style="text-align: center">$a_{10}$</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">$\beta_1~n_{G1}$</td>
      <td style="text-align: center">$\beta_2~n_{G2}$</td>
      <td style="text-align: center">$\gamma_1~n_{P_1}$</td>
      <td style="text-align: center">$\gamma_2~n_{P_2}$</td>
      <td style="text-align: center">$\kappa_{c1}^{+}n_{P2}n_{G1}$</td>
      <td style="text-align: center">$\kappa_{c1}^{-}~\xi_{1}^{I}$</td>
      <td style="text-align: center">$\kappa_{c2}^{+}n_{P1}n_{G2}$</td>
      <td style="text-align: center">$\kappa_{c2}^{-}~\xi_{2}^{I_1}$</td>
      <td style="text-align: center">$\kappa_{c3}^{+}n_{P1}\xi_{2}^{I_1}$</td>
      <td style="text-align: center">$\kappa_{c3}^{-}~\xi_{2}^{I_2}$</td>
    </tr>
  </tbody>
</table>

<p>As previously we assume, that DNA <a href="https://www.quora.com/Why-is-DNA-stable-but-RNA-is-not">does not degrade</a>, although the transcription from it can be inhibited by binding with respective protein (reactions $3$ and $4$).</p>

<p>Further we set the initial conditions as following:</p>

<table>
  <thead>
    <tr>
      <th style="text-align: center">$G_1$</th>
      <th style="text-align: center">$P_1$</th>
      <th style="text-align: center">$G_2$</th>
      <th style="text-align: center">$P_2$</th>
      <th style="text-align: center">$\xi_{1}^{I}$</th>
      <th style="text-align: center">$\xi_{2}^{I_1}$</th>
      <th style="text-align: center">$\xi_{2}^{I_2}$</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">1</td>
      <td style="text-align: center">10</td>
      <td style="text-align: center">1</td>
      <td style="text-align: center">10</td>
      <td style="text-align: center">0</td>
      <td style="text-align: center">0</td>
      <td style="text-align: center">0</td>
    </tr>
  </tbody>
</table>

<h3 id="simulation-1">Simulation</h3>
<p><br />
As the code in previous example was written with to be generalizable, we will be able to almost entirely reuse it, we just need to update the model parameters as follows:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Model Parameters
</span><span class="n">sigma</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">V</span> <span class="o">=</span> <span class="mf">1.5e-14</span>
<span class="n">to_nanomolar</span> <span class="o">=</span> <span class="mf">1e-9</span>
<span class="n">omega</span> <span class="o">=</span> <span class="mf">6.02214086e23</span><span class="o">*</span><span class="n">V</span><span class="o">*</span><span class="n">to_nanomolar</span>
<span class="n">t_end</span> <span class="o">=</span> <span class="mi">30000</span>
<span class="n">k</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">asarray</span><span class="p">([</span><span class="mi">100</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="o">/</span><span class="n">omega</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="o">/</span><span class="n">omega</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="o">*</span><span class="n">sigma</span><span class="o">/</span><span class="n">omega</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">q0</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">array</span><span class="p">([</span>  <span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">r</span><span class="s">'$G_1$'</span><span class="p">,</span> <span class="sa">r</span><span class="s">'$P_1$'</span><span class="p">,</span> <span class="sa">r</span><span class="s">'$G_2$'</span><span class="p">,</span> <span class="sa">r</span><span class="s">'$P_2$'</span><span class="p">,</span>
         <span class="sa">r</span><span class="s">'$\xi_1^{I}$'</span><span class="p">,</span> <span class="sa">r</span><span class="s">'$\xi_2^{I_1}$'</span><span class="p">,</span> <span class="sa">r</span><span class="s">'$\xi_2^{I_2}$'</span><span class="p">]</span>
<span class="n">dist</span> <span class="o">=</span> <span class="p">[</span><span class="s">"none"</span><span class="p">]</span><span class="o">*</span><span class="mi">7</span>
<span class="c1">## State change matrix associated with each reaction
</span><span class="n">nu</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">array</span><span class="p">([</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="c1">#G1    -&gt; G1+P1
</span>                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="c1">#G2    -&gt; G2+P2
</span>                <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="c1">#P1    -&gt; _
</span>                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="c1">#P2    -&gt; _
</span>                <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="c1">#G1+P2 -&gt; C1
</span>                <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="c1">#C1    -&gt; G1+P2
</span>                <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="c1">#P1+G2 -&gt; C2a
</span>                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="c1">#C2a   -&gt; P1+G2
</span>                <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="c1">#P1+C2a-&gt; C2b
</span>                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>  <span class="c1">#C2b   -&gt; P1+C2a
</span>              <span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="nb">int</span><span class="p">)</span>
<span class="c1">## Molecularity of species entering each reaction
</span><span class="n">psi</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="nb">int</span><span class="p">)</span>

<span class="n">mp_toggle</span> <span class="o">=</span> <span class="n">ssa</span><span class="p">.</span><span class="n">modelParameters</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">nu</span><span class="p">,</span> <span class="n">psi</span><span class="p">,</span> <span class="n">q0</span><span class="p">,</span> <span class="n">t_end</span><span class="p">,</span> <span class="n">names</span><span class="p">,</span> <span class="n">dist</span><span class="p">)</span>

<span class="c1">#Simulation Parameters
</span><span class="n">n_paths</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">do_load</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># default 0
</span><span class="n">do_save</span> <span class="o">=</span> <span class="mi">1</span> <span class="c1"># default 1
</span><span class="n">np</span><span class="p">.</span><span class="n">random</span><span class="p">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ssa</span><span class="p">.</span><span class="n">main</span><span class="p">(</span><span class="n">mp_toggle</span><span class="p">,</span> <span class="n">n_paths</span><span class="p">,</span> <span class="n">do_load</span> <span class="p">,</span><span class="n">do_save</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-html highlighter-rouge"><div class="highlight"><pre class="highlight"><code>2018-04-15 09:43:40,516:ssa_routine:INFO: Starting simulation with n_paths=1
2018-04-15 09:43:40,524:ssa_routine:INFO: Start of simulation for path # 1 out of 1.
2018-04-15 09:45:07,171:ssa_routine:INFO: Relative time: 0.20.
2018-04-15 09:46:01,920:ssa_routine:INFO: Relative time: 0.40.
2018-04-15 09:48:00,145:ssa_routine:INFO: Relative time: 0.60.
2018-04-15 09:49:25,028:ssa_routine:INFO: Relative time: 0.80.
2018-04-15 09:52:52,343:ssa_routine:INFO: Data saved to file results~/data/out_20180415-095212.
2018-04-15 09:52:52,605:ssa_routine:INFO: Calculating statistics for all species.
2018-04-15 09:53:29,373:ssa_routine:INFO: Simulation finished.
2018-04-15 09:54:03,614:ssa_routine:INFO: Program execution finished.
</code></pre></div></div>

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: center;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: center;">
      <th></th>
      <th>mean</th>
      <th>var</th>
      <th>CV</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>$G_1$</th>
      <td>0.678624</td>
      <td>0.218094</td>
      <td>0.688165</td>
    </tr>
    <tr>
      <th>$P_1$</th>
      <td>7.137714</td>
      <td>31.071102</td>
      <td>0.780943</td>
    </tr>
    <tr>
      <th>$G_2$</th>
      <td>0.320242</td>
      <td>0.217687</td>
      <td>1.456927</td>
    </tr>
    <tr>
      <th>$P_2$</th>
      <td>159.794211</td>
      <td>54399.408869</td>
      <td>1.459607</td>
    </tr>
    <tr>
      <th>$\xi_1^{I}$</th>
      <td>0.321376</td>
      <td>0.218094</td>
      <td>1.453140</td>
    </tr>
    <tr>
      <th>$\xi_2^{I_1}$</th>
      <td>0.005893</td>
      <td>0.005858</td>
      <td>12.988131</td>
    </tr>
    <tr>
      <th>$\xi_2^{I_2}$</th>
      <td>0.673865</td>
      <td>0.219771</td>
      <td>0.695685</td>
    </tr>
  </tbody>
</table>
</div>

<h3 id="results-1">Results</h3>
<p><br />
From the statistics, we may conclude that there is not much going on in our system, but look at the following plots of temporal evolution of protein copy numbers is more interesting. We see that the system manifests bistability when there are significant and  abrupt jumps in copy numbers of both proteins. This is the expected effect of noise-induced switching that is not tractable in deterministic model. The bistable behavior is due to the fact that the copy numbers of the proteins are dependent on presence or absence of complexes $\xi_1^{I}$, $\xi_2^{I_1}$, $\xi_2^{I_2}$ that in turn acquire copy numbers only in set $\{0,1\}$. Such binary state is significantly affected by stochastic events (recall the previous example with parking lot).</p>

<div class="img_row">
    <a href="/img/advanced_example_evolution_P1.png" target="_blank">
	<img class="col three" src="/img/advanced_example_evolution_P1.png" alt="Easy Example" title="Easy Example" /></a>
</div>

<div class="img_row">
    <a href="/img/advanced_example_evolution_P2.png" target="_blank">
	<img class="col three" src="/img/advanced_example_evolution_P2.png" alt="Easy Example" title="Easy Example" /></a>
</div>
<div class="col three caption">
Copy number evolution for proteins 1 and 2. Both species exhibit bistability in state space that is the best seen in the marginal distribution on the right panel. Although the copy numbers are mostly around 10 and 0, this can suddenly and randomly change to 0 and 500 for proteins 1 and 2 respectively.
</div>

<h2 id="conclusion">Conclusion</h2>
<p><br />
If you have made it all the way here, pat yourself on the back. It has been quite some math we have covered and it allows you to faithfully simulate dynamical systems as intriguing and complicated as genetic circuits. Now you know something about the phenomenon of bistability, which is a crucial element of building a toggle-switch, one of the elemental parts in synthetic-biology. You  also understand the important role of noise in gene expression and that the inherent stochasticity is not purely a complication in our design but can be efficiently exploited to drive patterns in molecular ‘decision-making’. In addition to <em>Stochastic Simulation Algorithm</em>, you are now also aware of other, more approximate, approaches for modeling biochemical (or other stochastic) systems like <em>Chemical Langevin</em> or <em>Fokker-Planck</em> equations that can allow you to inspect systems of larger size.</p>

<p>Last but not least, after inspecting the code in the <a href="https://github.com/martinholub/gillespie-notebook">github repository</a>, you will have seen examples of some advanced python capabilities including function decorators, just-in-time compilation and logging.</p>

<p>I hope reading this post was as enjoyable for you as much as it was for me to write it. If you feel like you want to find more about the exciting field of synthetic biology, you can start with <a href="https://www.youtube.com/watch?v=lNttxYdGHs4">this video</a>. In any case, stay tuned for upcoming posts on cool interdisciplinary and engineering topics.</p>

<hr />

<p><br /></p>
<h2 id="suplementary-notes">Suplementary Notes</h2>
<p><br /></p>
<h3 id="time-step-size-derivation-for-ssa">Time step size derivation for SSA</h3>

<p>Define $p(\tau, j |\mathbf{q}, t)d\tau$ as the probability that the reaction $j$ occurs in the time interval $[t + \tau, t + \tau + d\tau]$, then we have:</p>

\[p(\tau, j ~|~ \mathbf{q}, t)d\tau = P_0 (\tau ~|~ \mathbf{q}, t) \cdot a_j(\mathbf{q})d\tau,\]

<p>where $P_0 (\tau ~|~ \mathbf{q}, t)$ is the probability that <em>no</em> reaction fires on time interval $[t, \tau + t]$. We note that starting time $t$ is arbitrary and we may set it to zero.
Corresponding probability that <em>no</em> reaction occurs in time interval $[t + \tau, t + \tau + d\tau]$:</p>

\[P_0 (\tau + d\tau\ |\ \mathbf{q}, t) = P_0 (\tau\ |\ \mathbf{q}, t) \cdot (1-ad\tau),\]

<p>where $a = \sum_{i}^{M}a_i.$</p>

<p>It then immediately follows that:</p>

\[\frac{d P_0}{d \tau} (\tau\ |\ \mathbf{q}, t) = -P_0 (\tau\ |\ \mathbf{q}, t) \cdot a.\]

<p>Solving this differential equation with initial condition $P_0 (0\ | \mathbf{q}, t) = 1$ yields:</p>

\[P_0 (\tau\ | \mathbf{q}, t) = \mathrm{exp}(-a\tau).\]

<p>Going back to previous definition, this gives us:</p>

\[p(\tau, j\ |\ \mathbf{q}, t) =  \mathrm{exp}(-a\tau) \cdot a_j(\mathbf{q}).\]

<p>To obtain index of next reaction $j$ we <em>marginalize</em> over all $\tau$ and obtain:</p>

\[\int_{0}^{\infty}p(\tau, j\ |\ \mathbf{q}, t) d\tau = a_j(\mathbf{q}) \int_{0}^{\infty} \mathrm{exp}(-a\tau)d\tau = \frac{a_j(\mathbf{q})}{a}.\]

<p>Similarly, to find time that next reaction occurs, we integrate over all reactions $j$ and obtain:</p>

\[p(\tau\ |\ \mathbf{q}, t) =  \mathrm{exp}(-a\tau) \cdot a.\]

<p>The probability that the reaction occurs on time interval $[0, \Delta t]$ is then given by the <em>cumulative distribution function (CDF)</em>:</p>

\[\int_{0}^{\Delta t}p(\tau\ | \mathbf{q}, t) d\tau = a \int_{0}^{\Delta t} \mathrm{exp}(-a\tau)d\tau = 1 - \mathrm{exp}(-a\Delta t).\]

<p>The CDF is bounded on $[0, 1]$ and we can easily generate samples from corresponding PDF by inverse-sampling when we draw uniformly distributed numbers $\xi$ which finally yields:</p>

\[\Delta t = \frac{1}{a} \mathrm{ln}\bigg(\frac{1}{1 - \xi}\bigg),\]

<p>as stated before.</p>

<h3 id="optimizing-the-sampling-in-ssa">Optimizing the sampling in SSA</h3>
<p><br />
The computational burden of the direct method for SSA hits us pretty early. This can be troubling as it prolongs development cycles and limits number of species we can track. To alleviate this at least marginally, we can try to speed up our current code without making any conceptual changes to what is being computed and how (in contrast to more approximate approaches like <em>Chemical Langevin Equation(CLE)</em> and <em>Fokker-Planck Equation (FPE)</em>).</p>

<h4 id="use-profiler-to-find-where-the-most-time-is-spent">Use profiler to find where the most time is spent</h4>
<p><br />
First, we profile our code to see where the most time is being spend.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">%</span><span class="n">load_ext</span> <span class="n">line_profiler</span>
<span class="kn">import</span> <span class="nn">ssa_routine</span> <span class="k">as</span> <span class="n">ssa</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="o">%</span><span class="n">lprun</span> <span class="o">-</span><span class="n">T</span> <span class="n">profiler</span><span class="p">.</span><span class="n">out</span> <span class="o">-</span><span class="n">f</span> <span class="n">ssa</span><span class="p">.</span><span class="n">ssa_routine</span> <span class="n">ssa</span><span class="p">.</span><span class="n">ssa_routine</span><span class="p">(</span><span class="n">mp</span><span class="p">.</span><span class="n">k</span><span class="p">,</span> <span class="n">mp</span><span class="p">.</span><span class="n">t_end</span><span class="p">,</span> <span class="n">mp</span><span class="p">.</span><span class="n">q0</span><span class="p">,</span> <span class="n">mp</span><span class="p">.</span><span class="n">nu</span><span class="p">,</span> <span class="n">mp</span><span class="p">.</span><span class="n">psi</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-html highlighter-rouge"><div class="highlight"><pre class="highlight"><code>\*\*\* Profile printout saved to text file 'profiler.out'.

Timer unit: 4.46227e-07 s
Total time: 3.26368 s
Function: ssa_routine at line 92

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    92                                           def ssa_routine(k, t_end, q0, nu):

    119     20580        71930      3.5      1.0      while t <span class="nt">&lt;</span> <span class="nt">t_end:</span>
    <span class="err">120</span>     <span class="err">20579</span>      <span class="err">2325960</span>    <span class="err">113.0</span>     <span class="err">31.8</span>          <span class="na">propensities</span><span class="err">,</span> <span class="na">total_prop = </span><span class="s">get_propensities(k,</span> <span class="na">state</span><span class="err">)</span>

    <span class="err">125</span>                                                   <span class="na">#</span> <span class="na">Draw</span> <span class="na">random</span> <span class="na">samples</span> <span class="na">for</span> <span class="na">update</span> <span class="na">of</span> <span class="na">state</span> <span class="na">and</span> <span class="na">time</span> <span class="na">step</span>
    <span class="err">126</span>     <span class="err">20579</span>      <span class="err">3574266</span>    <span class="err">173.7</span>     <span class="err">48.9</span>          <span class="na">idx</span><span class="err">,</span> <span class="na">tau = </span><span class="s">draw_ssa(propensities,</span> <span class="na">total_prop</span><span class="err">)</span>

    <span class="err">141</span>         <span class="err">1</span>            <span class="err">3</span>      <span class="err">3.0</span>      <span class="err">0.0</span>      <span class="na">return</span> <span class="na">Q</span><span class="err">,</span> <span class="na">time</span>
</code></pre></div></div>

<h4 id="use-numba-to-speed-up-the-calculations">Use numba to speed up the calculations</h4>
<p><br />
We have found out that absolute majority of the time is spent in computing propensities and drawing samples for update with a respective reaction in given time interval. Unsurprisingly, as these occur on each and every time-step. Clearly, we can get the most ‘bang for the buck’ by making these two steps faster. To this end, we place them into individual functions and use <a href="https://numba.pydata.org/"><code class="language-html highlighter-rouge">numba</code>’s</a> <code class="language-html highlighter-rouge">@jit</code> decorator for just-in-time (JIT) compilation. This allows us to get about five-fold speed-up in calculations which I think is fair given the simplicity of implementation. This boils down to <a href="https://wiki.python.org/moin/PythonDecorators">decorating</a> the function as:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">@</span><span class="n">numba</span><span class="p">.</span><span class="n">jit</span><span class="p">(</span><span class="n">nopython</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">some_function</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">x</span> <span class="o">+</span> <span class="n">y</span>
</code></pre></div></div>

<p>Here is the speed comparison:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">%</span><span class="n">set_env</span> <span class="n">NUMBA_DISABLE_JIT</span><span class="o">=</span><span class="mi">1</span>
<span class="c1">#os.environ["NUMBA_DISABLE_JIT"] = "1"
</span><span class="o">%</span><span class="n">timeit</span> <span class="o">-</span><span class="n">n</span> <span class="mi">1</span> <span class="o">-</span><span class="n">r</span> <span class="mi">3</span> <span class="n">ssa</span><span class="p">.</span><span class="n">simulate</span><span class="p">(</span><span class="n">mp</span><span class="p">,</span> <span class="n">n_paths</span><span class="p">)</span>
</code></pre></div></div>

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean</th>
      <th>var</th>
      <th>CV</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>mRNA</th>
      <td>10.394067</td>
      <td>9.208149</td>
      <td>3.425306</td>
    </tr>
    <tr>
      <th>Protein</th>
      <td>249.466942</td>
      <td>1694.715664</td>
      <td>6.059888</td>
    </tr>
  </tbody>
</table>
</div>

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean</th>
      <th>var</th>
      <th>CV</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>mRNA</th>
      <td>10.688117</td>
      <td>9.27994</td>
      <td>3.508558</td>
    </tr>
    <tr>
      <th>Protein</th>
      <td>257.408644</td>
      <td>1726.12629</td>
      <td>6.195650</td>
    </tr>
  </tbody>
</table>
</div>

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean</th>
      <th>var</th>
      <th>CV</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>mRNA</th>
      <td>10.467691</td>
      <td>8.531862</td>
      <td>3.583678</td>
    </tr>
    <tr>
      <th>Protein</th>
      <td>255.014802</td>
      <td>1703.876616</td>
      <td>6.177977</td>
    </tr>
  </tbody>
</table>
</div>

<div class="language-html highlighter-rouge"><div class="highlight"><pre class="highlight"><code>13.7 s ± 225 ms per loop (mean ± std. dev. of 3 runs, 1 loop each)
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">%</span><span class="n">set_env</span> <span class="n">NUMBA_DISABLE_JIT</span><span class="o">=</span><span class="mi">0</span>
<span class="c1">#os.environ["NUMBA_DISABLE_JIT"] = "0"
</span><span class="o">%</span><span class="n">timeit</span> <span class="o">-</span><span class="n">n</span> <span class="mi">1</span> <span class="o">-</span><span class="n">r</span> <span class="mi">3</span> <span class="n">ssa</span><span class="p">.</span><span class="n">simulate</span><span class="p">(</span><span class="n">mp</span><span class="p">,</span> <span class="n">n_paths</span><span class="p">)</span>
</code></pre></div></div>

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean</th>
      <th>var</th>
      <th>CV</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>gene</th>
      <td>1.000000</td>
      <td>0.000000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>mRNA</th>
      <td>10.999584</td>
      <td>11.375058</td>
      <td>3.261365</td>
    </tr>
    <tr>
      <th>Protein</th>
      <td>263.212386</td>
      <td>2612.338753</td>
      <td>5.149814</td>
    </tr>
  </tbody>
</table>
</div>

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean</th>
      <th>var</th>
      <th>CV</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>gene</th>
      <td>1.000000</td>
      <td>0.000000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>mRNA</th>
      <td>10.532080</td>
      <td>10.249184</td>
      <td>3.289800</td>
    </tr>
    <tr>
      <th>Protein</th>
      <td>253.934811</td>
      <td>1920.826411</td>
      <td>5.793996</td>
    </tr>
  </tbody>
</table>
</div>

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean</th>
      <th>var</th>
      <th>CV</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>gene</th>
      <td>1.000000</td>
      <td>0.000000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>mRNA</th>
      <td>10.366590</td>
      <td>9.328903</td>
      <td>3.394069</td>
    </tr>
    <tr>
      <th>Protein</th>
      <td>249.273439</td>
      <td>1465.737686</td>
      <td>6.511003</td>
    </tr>
  </tbody>
</table>
</div>

<div class="language-html highlighter-rouge"><div class="highlight"><pre class="highlight"><code>2.41 s ± 580 ms per loop (mean ± std. dev. of 3 runs, 1 loop each)
</code></pre></div></div>

  </article>

  
<!---
Call Fontawesome in the head section or in the location where you place the share bar
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
--->
<hr>
<br/>
<!---<center><p><script async src="https://eocampaign1.com/form/cfe45fdc-109a-11ef-a9e2-bf0899395051.js" data-form="cfe45fdc-109a-11ef-a9e2-bf0899395051"></script></p></center>--->
<div class="share-box center" style="width: 60%; margin: 0px auto;">
<h2 class="center">Share this:</h2>
<br/>
<a class="l" href="https://www.linkedin.com/shareArticle?mini=true&url=/2018/04/15/stochsim.html&title=Stochastic Modeling in Biology&summary=&source=martinholub" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fa-brands fa-linkedin"></i><span> linkedin</span></a>

<a class="t" href="https://twitter.com/intent/tweet?text=Stochastic Modeling in Biology&url=/2018/04/15/stochsim.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fa-brands fa-square-x-twitter"></i><span> x</span></a>

<a class="t" href="https://bsky.app/intent/post?text=Stochastic Modeling in Biology&url=/2018/04/15/stochsim.html" onclick="window.open(this.href, 'mywin', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;">
  <i class="fa-brands fa-square-bluesky"></i><span> bluesky</span>
</a>
<!---
<a class="g" href="https://plus.google.com/share?url=/2018/04/15/stochsim.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" ><i class="fa fa-google-plus fa"></i><span> google</span></a>
--->
<a class="r" href="http://www.reddit.com/submit?url=/2018/04/15/stochsim.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=900,height=500,toolbar=1,resizable=0'); return false;"><i class="fa-brands fa-square-reddit"></i><span> reddit</span></a>

<a class="f" href="https://www.facebook.com/sharer/sharer.php?u=/2018/04/15/stochsim.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" ><i class="fa-brands fa-square-facebook"></i><span> facebook</span></a>

<a class="e" href="mailto:?subject=Stochastic Modeling in Biology&amp;body=Check out this site /2018/04/15/stochsim.html"><i class="fa fa-envelope fa"></i><span> email</span></a>
</div>
<br/>
<hr>


  

<div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'martinholub'; // required: replace example with your forum shortname
        /*var disqus_developer = 1; // Comment out when the site is live */
        var disqus_identifier = "/2018/04/15/stochsim.html";

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>




</div>

      </div>
    </div>

<!--
    <span class="contacticon center">
  <a href="https://www.instagram.com/scimartin/" target="_blank"><i class="fa-brands fa-square-instagram"></i></a>
  <a href="https://www.linkedin.com/in/scimartin" target="_blank"><i class="fa-brands fa-linkedin"></i></a>
  <a href="https://bsky.app/profile/scimartin.bsky.social" target="_blank"><i class="fa-brands fa-square-bluesky"></i></a>
  <div class="iconsuperwrapper">
  <a href="https://paragraph.xyz/@scimartin" target="_blank">
  <div class="iconwrapper">
  <svg
   class="icon"
   version="1.0"
   width="36"
   height="36"
   viewBox="0 0 36 36"
   preserveAspectRatio="xMidYMid meet"
   id="svg10"
   sodipodi:docname="paragraph.svg"
   inkscape:version="1.1.1 (3bf5ae0d25, 2021-09-20)"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:svg="http://www.w3.org/2000/svg">
  <path
     id="path4"
     d="M 0 0 L 0 12.7273 L 0 25.4546 L 3.73818 25.4546 C 6.04363 25.4546 7.49818 25.4255 7.52727 25.3819 C 7.54909 25.3455 7.61454 25.331 7.66545 25.3528 C 7.82545 25.411 9.96371 25.4254 9.96371 25.3673 C 9.96371 25.3382 10.0073 25.331 10.0582 25.3528 C 10.1164 25.3746 10.2546 25.3964 10.371 25.4037 C 10.5383 25.4182 10.5819 25.4545 10.5746 25.5854 C 10.5746 25.6727 10.5964 25.7455 10.6255 25.7455 C 10.6546 25.7455 10.6546 25.84 10.6183 25.9564 C 10.5892 26.0655 10.5892 26.2036 10.6183 26.2618 C 10.6547 26.3127 10.6547 26.4073 10.6183 26.4728 C 10.5819 26.5382 10.5819 26.6327 10.6109 26.6764 C 10.64 26.7273 10.6473 26.9309 10.6183 27.1273 C 10.5892 27.3237 10.5964 27.5274 10.6255 27.5783 C 10.6546 27.6219 10.6546 27.7164 10.6183 27.7819 C 10.5819 27.8473 10.5819 27.9419 10.6183 28 C 10.6547 28.0582 10.6547 28.1528 10.6183 28.2182 C 10.5819 28.2837 10.5819 28.3782 10.6183 28.4364 C 10.6474 28.4873 10.6474 28.611 10.6183 28.6983 C 10.5747 28.8001 10.5819 28.8873 10.6183 28.9091 C 10.6547 28.9309 10.6547 29.0401 10.6183 29.1637 C 10.5819 29.2873 10.5819 29.3819 10.6183 29.3819 C 10.6547 29.3819 10.6547 29.4764 10.6183 29.6001 C 10.5819 29.7237 10.5819 29.8183 10.6183 29.8183 C 10.6547 29.8183 10.6547 29.9128 10.6183 30.0364 C 10.5819 30.1528 10.5746 30.2546 10.6037 30.2546 C 10.6328 30.2546 10.6546 30.3709 10.6546 30.5091 C 10.6546 30.6473 10.6328 30.7636 10.6037 30.7636 C 10.5746 30.7636 10.5819 30.8582 10.6183 30.9818 C 10.6547 31.0982 10.6547 31.2 10.6183 31.2 C 10.5892 31.2 10.5819 31.2874 10.6109 31.3965 C 10.64 31.5128 10.6401 31.7164 10.6183 31.8691 C 10.5892 32.0146 10.5892 32.1746 10.6255 32.2255 C 10.6546 32.2764 10.6546 32.371 10.6183 32.4364 C 10.5819 32.5019 10.5819 32.5964 10.6183 32.6546 C 10.6547 32.7128 10.6547 32.8073 10.6183 32.8728 C 10.5819 32.9383 10.5819 33.0328 10.6183 33.091 C 10.6474 33.1419 10.6474 33.2655 10.6183 33.3527 C 10.5747 33.4546 10.5819 33.5419 10.6183 33.5637 C 10.6547 33.5855 10.6547 33.6945 10.6183 33.8182 C 10.5819 33.9418 10.5819 34.0364 10.6183 34.0364 C 10.6547 34.0364 10.6547 34.1309 10.6183 34.2545 C 10.5819 34.3782 10.5819 34.4727 10.6183 34.4727 C 10.6547 34.4727 10.6547 34.5673 10.6183 34.6909 C 10.5819 34.8073 10.5746 34.9091 10.6037 34.9091 C 10.6619 34.9091 10.6546 35.4037 10.5965 35.4619 C 10.5747 35.4837 10.582 35.5782 10.6183 35.6727 C 10.6547 35.7673 10.6619 35.8618 10.6401 35.8837 C 10.611 35.9055 16.3128 35.9273 23.2946 35.9273 L 36 35.9273 L 36 22.5455 L 36 9.16364 L 34.7273 9.16364 L 33.4546 9.16364 L 33.4546 21.2873 L 33.4546 33.4183 L 28.5673 33.3964 C 25.8764 33.3891 22.9382 33.3819 22.0364 33.3892 C 21.1345 33.3892 20.3491 33.3964 20.2909 33.3964 C 20.2327 33.3891 18.6327 33.3892 16.7491 33.3892 L 13.3091 33.3818 L 13.3091 29.3819 L 13.3091 25.3746 L 13.5491 25.3818 C 13.6727 25.3818 13.9492 25.3818 14.1601 25.3746 C 14.371 25.3746 14.5746 25.3891 14.6109 25.4109 C 14.6473 25.44 14.6982 25.4255 14.7273 25.3819 C 14.7564 25.3382 14.8074 25.3237 14.8437 25.3528 C 14.9092 25.3891 17.5564 25.4037 18.1091 25.36 C 18.2473 25.3527 18.4364 25.3673 18.5237 25.3964 C 18.611 25.4255 18.7055 25.4182 18.7273 25.3746 C 18.7564 25.3382 18.8073 25.3236 18.8436 25.3527 C 18.8873 25.3745 19.091 25.3963 19.2946 25.3891 C 19.5055 25.3818 19.7455 25.3819 19.8401 25.3819 C 20.1746 25.3819 21.4255 25.1637 21.9273 25.0183 C 22.4364 24.8655 22.56 24.8364 22.7637 24.7855 C 22.8219 24.7782 22.8873 24.7491 22.9091 24.7273 C 22.9309 24.7055 23.2073 24.56 23.5201 24.4073 C 24.1382 24.1164 24.7782 23.7019 25.1127 23.3891 C 25.2218 23.2873 25.3382 23.2001 25.3673 23.1928 C 25.4545 23.1855 26.2037 22.2619 26.5018 21.7964 C 26.8146 21.3164 27.3382 20.0655 27.3455 19.811 C 27.3455 19.731 27.3673 19.6364 27.3964 19.6073 C 27.4255 19.5782 27.4546 19.4327 27.4618 19.28 C 27.4618 19.1273 27.491 18.9601 27.5273 18.9092 C 27.6437 18.7201 27.5928 16.371 27.4546 15.7455 C 27.3891 15.4255 27.3237 15.1127 27.3092 15.0545 C 27.2946 14.96 27.2073 14.7127 27.0764 14.4 C 27.0546 14.3418 27.0255 14.2546 27.0255 14.2182 C 27.0036 14.0582 26.5382 13.2437 26.2255 12.8292 C 25.2219 11.4837 23.3019 10.2764 21.6364 9.92003 C 20.3346 9.64366 20.3564 9.65093 15.4473 9.62184 L 10.6327 9.59275 L 10.6037 9.81093 C 10.5818 9.93457 10.5964 10.0364 10.6255 10.0364 C 10.6546 10.0364 10.6546 10.1309 10.6183 10.2545 C 10.5819 10.3782 10.5819 10.4727 10.6183 10.4727 C 10.6547 10.4727 10.6547 10.5673 10.6183 10.6909 C 10.5819 10.8145 10.5819 10.9091 10.6183 10.9091 C 10.6547 10.9091 10.6547 11.0037 10.6183 11.12 C 10.5892 11.2291 10.5892 11.3673 10.6183 11.4255 C 10.6547 11.4764 10.6547 11.5709 10.6183 11.6364 C 10.5819 11.7019 10.5819 11.7964 10.6183 11.8546 C 10.6547 11.9128 10.6547 12.0073 10.6183 12.0728 C 10.5819 12.1382 10.5819 12.2328 10.6183 12.2909 C 10.6547 12.3491 10.6547 12.4436 10.6183 12.5164 C 10.5819 12.5818 10.5674 12.6546 10.5965 12.6764 C 10.6474 12.7346 10.6546 13.4473 10.5965 13.4982 C 10.5746 13.52 10.5892 13.6364 10.6183 13.7528 C 10.6547 13.8692 10.6547 13.9782 10.6183 14.0001 C 10.5892 14.0219 10.5892 14.1455 10.6183 14.2837 C 10.6474 14.4218 10.6474 14.5745 10.6183 14.6182 C 10.5892 14.6691 10.5892 14.8 10.6183 14.9164 C 10.6547 15.0255 10.6546 15.1345 10.6328 15.1636 C 10.5746 15.2145 10.5965 15.9055 10.6546 16 C 10.6765 16.0437 10.6692 16.0728 10.6256 16.0728 C 10.5819 16.0728 10.5819 16.1528 10.6184 16.291 C 10.6547 16.4146 10.6547 16.5092 10.6184 16.5092 C 10.582 16.5092 10.582 16.6037 10.6184 16.7273 C 10.6547 16.851 10.6547 16.9455 10.6184 16.9455 C 10.582 16.9455 10.582 17.0401 10.6184 17.1637 C 10.6547 17.2873 10.6547 17.3819 10.6184 17.3819 C 10.582 17.3819 10.582 17.4764 10.6184 17.6001 C 10.6547 17.7237 10.6547 17.8182 10.6184 17.8182 C 10.582 17.8182 10.582 17.9128 10.6184 18.0364 C 10.6547 18.1601 10.6547 18.2546 10.6184 18.2546 C 10.582 18.2546 10.582 18.3492 10.6184 18.4728 C 10.6547 18.5964 10.6547 18.691 10.6184 18.691 C 10.582 18.691 10.582 18.7855 10.6184 18.9092 C 10.6547 19.0328 10.6547 19.1273 10.6184 19.1273 C 10.582 19.1273 10.582 19.2219 10.6184 19.3455 C 10.6547 19.4692 10.6547 19.5637 10.6184 19.5637 C 10.582 19.5637 10.5675 19.6073 10.5893 19.6654 C 10.6547 19.84 10.6619 20.24 10.6038 20.4146 C 10.531 20.64 10.6983 20.7419 11.142 20.7346 C 12.0438 20.7128 12.2184 20.72 12.2838 20.7563 C 12.3202 20.7854 12.3711 20.771 12.3929 20.7346 C 12.422 20.6983 12.5238 20.6909 12.6256 20.72 C 12.7347 20.7491 12.931 20.7564 13.0619 20.7419 L 13.3092 20.7128 L 13.2947 16.3855 L 13.2729 12.0582 L 16.6547 12.0946 C 19.9492 12.131 20.5965 12.1745 20.9311 12.3854 C 20.9747 12.4218 21.0183 12.4291 21.0183 12.4073 C 21.0183 12.3346 21.8474 12.6619 22.2474 12.8873 C 23.5492 13.6146 24.2984 14.5819 24.662 16 C 24.8656 16.7782 24.8583 18.2909 24.6546 19.0837 C 24.4582 19.8618 24.0583 20.6546 23.6292 21.1055 C 23.4328 21.3164 23.2729 21.5127 23.2729 21.5418 C 23.2729 21.5709 23.2292 21.6 23.1783 21.6 C 23.1273 21.6 23.0182 21.6582 22.9382 21.7309 C 22.8 21.8473 22.7055 21.9055 22.2255 22.1745 C 21.8764 22.3709 21.1855 22.6182 20.5091 22.7855 C 19.9055 22.931 19.7165 22.9382 11.2219 22.96 L 2.54558 22.9891 L 2.54558 12.7346 L 2.54558 2.47276 L 14.6547 2.47276 L 26.7638 2.47276 L 26.7638 1.23643 L 26.7638 0 L 13.3819 0 Z M 29.3819 0 L 29.3819 1.23643 L 29.3819 2.47276 L 31.4182 2.47276 L 33.4546 2.47276 L 33.4546 4.50916 L 33.4546 6.54545 L 34.7273 6.54545 L 36 6.54545 L 36 3.27273 L 36 0 L 32.6909 0 Z" />
  </svg>
  </div>
  </a>
  </div>
  <br>
  <a href="mailto:mholub.ethz=gmail+com"><i class="fa fa-envelope-square"></i></a>
  <a href="https://orcid.org/0000-0002-8365-0927" target="_blank"><i class="ai ai-orcid"></i></a>
  <a href="https://github.com/martinholub" target="_blank"><i class="fa-brands fa-square-github"></i></a>
  <a href="/feed.xml" target="_blank"><i class="fa fa-rss-square"></i></a>
</span>


<div class="col three caption">
	&#169; Martin Holub, 2025
</div>
-->
  <script id="dsq-count-scr" src="//martinholub.disqus.com/count.js" async></script>
  <script src="/css/lightbox.js"></script>

  </body>
    <span class="contacticon center">
  <a href="https://www.instagram.com/scimartin/" target="_blank"><i class="fa-brands fa-square-instagram"></i></a>
  <a href="https://www.linkedin.com/in/scimartin" target="_blank"><i class="fa-brands fa-linkedin"></i></a>
  <a href="https://bsky.app/profile/scimartin.bsky.social" target="_blank"><i class="fa-brands fa-square-bluesky"></i></a>
  <div class="iconsuperwrapper">
  <a href="https://paragraph.xyz/@scimartin" target="_blank">
  <div class="iconwrapper">
  <svg
   class="icon"
   version="1.0"
   width="36"
   height="36"
   viewBox="0 0 36 36"
   preserveAspectRatio="xMidYMid meet"
   id="svg10"
   sodipodi:docname="paragraph.svg"
   inkscape:version="1.1.1 (3bf5ae0d25, 2021-09-20)"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:svg="http://www.w3.org/2000/svg">
  <path
     id="path4"
     d="M 0 0 L 0 12.7273 L 0 25.4546 L 3.73818 25.4546 C 6.04363 25.4546 7.49818 25.4255 7.52727 25.3819 C 7.54909 25.3455 7.61454 25.331 7.66545 25.3528 C 7.82545 25.411 9.96371 25.4254 9.96371 25.3673 C 9.96371 25.3382 10.0073 25.331 10.0582 25.3528 C 10.1164 25.3746 10.2546 25.3964 10.371 25.4037 C 10.5383 25.4182 10.5819 25.4545 10.5746 25.5854 C 10.5746 25.6727 10.5964 25.7455 10.6255 25.7455 C 10.6546 25.7455 10.6546 25.84 10.6183 25.9564 C 10.5892 26.0655 10.5892 26.2036 10.6183 26.2618 C 10.6547 26.3127 10.6547 26.4073 10.6183 26.4728 C 10.5819 26.5382 10.5819 26.6327 10.6109 26.6764 C 10.64 26.7273 10.6473 26.9309 10.6183 27.1273 C 10.5892 27.3237 10.5964 27.5274 10.6255 27.5783 C 10.6546 27.6219 10.6546 27.7164 10.6183 27.7819 C 10.5819 27.8473 10.5819 27.9419 10.6183 28 C 10.6547 28.0582 10.6547 28.1528 10.6183 28.2182 C 10.5819 28.2837 10.5819 28.3782 10.6183 28.4364 C 10.6474 28.4873 10.6474 28.611 10.6183 28.6983 C 10.5747 28.8001 10.5819 28.8873 10.6183 28.9091 C 10.6547 28.9309 10.6547 29.0401 10.6183 29.1637 C 10.5819 29.2873 10.5819 29.3819 10.6183 29.3819 C 10.6547 29.3819 10.6547 29.4764 10.6183 29.6001 C 10.5819 29.7237 10.5819 29.8183 10.6183 29.8183 C 10.6547 29.8183 10.6547 29.9128 10.6183 30.0364 C 10.5819 30.1528 10.5746 30.2546 10.6037 30.2546 C 10.6328 30.2546 10.6546 30.3709 10.6546 30.5091 C 10.6546 30.6473 10.6328 30.7636 10.6037 30.7636 C 10.5746 30.7636 10.5819 30.8582 10.6183 30.9818 C 10.6547 31.0982 10.6547 31.2 10.6183 31.2 C 10.5892 31.2 10.5819 31.2874 10.6109 31.3965 C 10.64 31.5128 10.6401 31.7164 10.6183 31.8691 C 10.5892 32.0146 10.5892 32.1746 10.6255 32.2255 C 10.6546 32.2764 10.6546 32.371 10.6183 32.4364 C 10.5819 32.5019 10.5819 32.5964 10.6183 32.6546 C 10.6547 32.7128 10.6547 32.8073 10.6183 32.8728 C 10.5819 32.9383 10.5819 33.0328 10.6183 33.091 C 10.6474 33.1419 10.6474 33.2655 10.6183 33.3527 C 10.5747 33.4546 10.5819 33.5419 10.6183 33.5637 C 10.6547 33.5855 10.6547 33.6945 10.6183 33.8182 C 10.5819 33.9418 10.5819 34.0364 10.6183 34.0364 C 10.6547 34.0364 10.6547 34.1309 10.6183 34.2545 C 10.5819 34.3782 10.5819 34.4727 10.6183 34.4727 C 10.6547 34.4727 10.6547 34.5673 10.6183 34.6909 C 10.5819 34.8073 10.5746 34.9091 10.6037 34.9091 C 10.6619 34.9091 10.6546 35.4037 10.5965 35.4619 C 10.5747 35.4837 10.582 35.5782 10.6183 35.6727 C 10.6547 35.7673 10.6619 35.8618 10.6401 35.8837 C 10.611 35.9055 16.3128 35.9273 23.2946 35.9273 L 36 35.9273 L 36 22.5455 L 36 9.16364 L 34.7273 9.16364 L 33.4546 9.16364 L 33.4546 21.2873 L 33.4546 33.4183 L 28.5673 33.3964 C 25.8764 33.3891 22.9382 33.3819 22.0364 33.3892 C 21.1345 33.3892 20.3491 33.3964 20.2909 33.3964 C 20.2327 33.3891 18.6327 33.3892 16.7491 33.3892 L 13.3091 33.3818 L 13.3091 29.3819 L 13.3091 25.3746 L 13.5491 25.3818 C 13.6727 25.3818 13.9492 25.3818 14.1601 25.3746 C 14.371 25.3746 14.5746 25.3891 14.6109 25.4109 C 14.6473 25.44 14.6982 25.4255 14.7273 25.3819 C 14.7564 25.3382 14.8074 25.3237 14.8437 25.3528 C 14.9092 25.3891 17.5564 25.4037 18.1091 25.36 C 18.2473 25.3527 18.4364 25.3673 18.5237 25.3964 C 18.611 25.4255 18.7055 25.4182 18.7273 25.3746 C 18.7564 25.3382 18.8073 25.3236 18.8436 25.3527 C 18.8873 25.3745 19.091 25.3963 19.2946 25.3891 C 19.5055 25.3818 19.7455 25.3819 19.8401 25.3819 C 20.1746 25.3819 21.4255 25.1637 21.9273 25.0183 C 22.4364 24.8655 22.56 24.8364 22.7637 24.7855 C 22.8219 24.7782 22.8873 24.7491 22.9091 24.7273 C 22.9309 24.7055 23.2073 24.56 23.5201 24.4073 C 24.1382 24.1164 24.7782 23.7019 25.1127 23.3891 C 25.2218 23.2873 25.3382 23.2001 25.3673 23.1928 C 25.4545 23.1855 26.2037 22.2619 26.5018 21.7964 C 26.8146 21.3164 27.3382 20.0655 27.3455 19.811 C 27.3455 19.731 27.3673 19.6364 27.3964 19.6073 C 27.4255 19.5782 27.4546 19.4327 27.4618 19.28 C 27.4618 19.1273 27.491 18.9601 27.5273 18.9092 C 27.6437 18.7201 27.5928 16.371 27.4546 15.7455 C 27.3891 15.4255 27.3237 15.1127 27.3092 15.0545 C 27.2946 14.96 27.2073 14.7127 27.0764 14.4 C 27.0546 14.3418 27.0255 14.2546 27.0255 14.2182 C 27.0036 14.0582 26.5382 13.2437 26.2255 12.8292 C 25.2219 11.4837 23.3019 10.2764 21.6364 9.92003 C 20.3346 9.64366 20.3564 9.65093 15.4473 9.62184 L 10.6327 9.59275 L 10.6037 9.81093 C 10.5818 9.93457 10.5964 10.0364 10.6255 10.0364 C 10.6546 10.0364 10.6546 10.1309 10.6183 10.2545 C 10.5819 10.3782 10.5819 10.4727 10.6183 10.4727 C 10.6547 10.4727 10.6547 10.5673 10.6183 10.6909 C 10.5819 10.8145 10.5819 10.9091 10.6183 10.9091 C 10.6547 10.9091 10.6547 11.0037 10.6183 11.12 C 10.5892 11.2291 10.5892 11.3673 10.6183 11.4255 C 10.6547 11.4764 10.6547 11.5709 10.6183 11.6364 C 10.5819 11.7019 10.5819 11.7964 10.6183 11.8546 C 10.6547 11.9128 10.6547 12.0073 10.6183 12.0728 C 10.5819 12.1382 10.5819 12.2328 10.6183 12.2909 C 10.6547 12.3491 10.6547 12.4436 10.6183 12.5164 C 10.5819 12.5818 10.5674 12.6546 10.5965 12.6764 C 10.6474 12.7346 10.6546 13.4473 10.5965 13.4982 C 10.5746 13.52 10.5892 13.6364 10.6183 13.7528 C 10.6547 13.8692 10.6547 13.9782 10.6183 14.0001 C 10.5892 14.0219 10.5892 14.1455 10.6183 14.2837 C 10.6474 14.4218 10.6474 14.5745 10.6183 14.6182 C 10.5892 14.6691 10.5892 14.8 10.6183 14.9164 C 10.6547 15.0255 10.6546 15.1345 10.6328 15.1636 C 10.5746 15.2145 10.5965 15.9055 10.6546 16 C 10.6765 16.0437 10.6692 16.0728 10.6256 16.0728 C 10.5819 16.0728 10.5819 16.1528 10.6184 16.291 C 10.6547 16.4146 10.6547 16.5092 10.6184 16.5092 C 10.582 16.5092 10.582 16.6037 10.6184 16.7273 C 10.6547 16.851 10.6547 16.9455 10.6184 16.9455 C 10.582 16.9455 10.582 17.0401 10.6184 17.1637 C 10.6547 17.2873 10.6547 17.3819 10.6184 17.3819 C 10.582 17.3819 10.582 17.4764 10.6184 17.6001 C 10.6547 17.7237 10.6547 17.8182 10.6184 17.8182 C 10.582 17.8182 10.582 17.9128 10.6184 18.0364 C 10.6547 18.1601 10.6547 18.2546 10.6184 18.2546 C 10.582 18.2546 10.582 18.3492 10.6184 18.4728 C 10.6547 18.5964 10.6547 18.691 10.6184 18.691 C 10.582 18.691 10.582 18.7855 10.6184 18.9092 C 10.6547 19.0328 10.6547 19.1273 10.6184 19.1273 C 10.582 19.1273 10.582 19.2219 10.6184 19.3455 C 10.6547 19.4692 10.6547 19.5637 10.6184 19.5637 C 10.582 19.5637 10.5675 19.6073 10.5893 19.6654 C 10.6547 19.84 10.6619 20.24 10.6038 20.4146 C 10.531 20.64 10.6983 20.7419 11.142 20.7346 C 12.0438 20.7128 12.2184 20.72 12.2838 20.7563 C 12.3202 20.7854 12.3711 20.771 12.3929 20.7346 C 12.422 20.6983 12.5238 20.6909 12.6256 20.72 C 12.7347 20.7491 12.931 20.7564 13.0619 20.7419 L 13.3092 20.7128 L 13.2947 16.3855 L 13.2729 12.0582 L 16.6547 12.0946 C 19.9492 12.131 20.5965 12.1745 20.9311 12.3854 C 20.9747 12.4218 21.0183 12.4291 21.0183 12.4073 C 21.0183 12.3346 21.8474 12.6619 22.2474 12.8873 C 23.5492 13.6146 24.2984 14.5819 24.662 16 C 24.8656 16.7782 24.8583 18.2909 24.6546 19.0837 C 24.4582 19.8618 24.0583 20.6546 23.6292 21.1055 C 23.4328 21.3164 23.2729 21.5127 23.2729 21.5418 C 23.2729 21.5709 23.2292 21.6 23.1783 21.6 C 23.1273 21.6 23.0182 21.6582 22.9382 21.7309 C 22.8 21.8473 22.7055 21.9055 22.2255 22.1745 C 21.8764 22.3709 21.1855 22.6182 20.5091 22.7855 C 19.9055 22.931 19.7165 22.9382 11.2219 22.96 L 2.54558 22.9891 L 2.54558 12.7346 L 2.54558 2.47276 L 14.6547 2.47276 L 26.7638 2.47276 L 26.7638 1.23643 L 26.7638 0 L 13.3819 0 Z M 29.3819 0 L 29.3819 1.23643 L 29.3819 2.47276 L 31.4182 2.47276 L 33.4546 2.47276 L 33.4546 4.50916 L 33.4546 6.54545 L 34.7273 6.54545 L 36 6.54545 L 36 3.27273 L 36 0 L 32.6909 0 Z" />
  </svg>
  </div>
  </a>
  </div>
  <br>
  <a href="mailto:mholub.ethz=gmail+com"><i class="fa fa-envelope-square"></i></a>
  <a href="https://orcid.org/0000-0002-8365-0927" target="_blank"><i class="ai ai-orcid"></i></a>
  <a href="https://github.com/martinholub" target="_blank"><i class="fa-brands fa-square-github"></i></a>
  <a href="/feed.xml" target="_blank"><i class="fa fa-rss-square"></i></a>
</span>


<div class="col three caption">
	&#169; Martin Holub, 2025
</div>

</html>
