<!DOCTYPE html>
<html>

  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width initial-scale=1" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <title>Using Deep Learning Libraries to Solve Stochastic Differential Equations</title>
    <meta name="description" content="Personal Website and Blog
">

    <link rel="stylesheet" href="/css/main.css">
    <link rel="canonical" href="https://www.martinholub.com/2018/05/01/TheanoTensorflowSDE.html">

    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.7.2/css/all.min.css">
	<link rel="stylesheet" href="/css/academicons.css"/>
    <link href="/css/lightbox.css" rel="stylesheet">

    <link rel="stylesheet" href="/css/zenburn.css"/>

    <!-- collect tags -->
    
      <!-- collect tags from blog posts -->
<!-- source: https://longqian.me/2017/02/09/github-jekyll-tag/ --->







    

	<link type="application/atom+xml" rel="alternate" href="https://www.martinholub.com/feed.xml" title="Martin Holub Blog" />
</head>

  
  <!-- Google tag (gtag.js) -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=G-1ZBQKKLMRB"></script>
  <script>
	window.dataLayer = window.dataLayer || [];
	function gtag(){dataLayer.push(arguments);}
	gtag('js', new Date());

	gtag('config', 'G-1ZBQKKLMRB');
  </script>

  <script type="text/javascript"
    src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
	</script>

  <body>

    <header class="site-header">

  <div class="wrapper">

    <nav class="site-nav">

      <div class="trigger">
        <!-- Martin Holub Blog instead of blog -->
        <a class="page-link" href="/">about</a>

        <!-- ( for page in site.pages )  was previously-->
        
          
          <a class="page-link" href="/blog">blog</a>
          
        
          
          <a class="page-link" href="/research">research</a>
          
        
          
          <a class="page-link" href="/projects">projects</a>
          
        
          
          <a class="page-link" href="/resources">resources</a>
          
        
          
          <a class="page-link" href="/subscribe">subscribe</a>
          
        

      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <header class="post-header">
    <h1 class="post-title">Using Deep Learning Libraries to Solve Stochastic Differential Equations</h1>
    <p class="post-meta">May 1, 2018 — 16:42 • <a href="https://martinholub.github.io/2018/05/01/TheanoTensorflowSDE.html#disqus_thread">0 Comments</a></p>
    <span class="post-meta">[
      
        
          <a href="/tag/ethz"><code class="highligher-rouge"><nobr>ethz</nobr></code>&nbsp;</a>
      
        
          <a href="/tag/projects"><code class="highligher-rouge"><nobr>projects</nobr></code>&nbsp;</a>
      
        
          <a href="/tag/code"><code class="highligher-rouge"><nobr>code</nobr></code>&nbsp;</a>
      
    ]</span>
  </header>
  
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["\\(","\\)"] ],
      displayMath: [ ['$$','$$'] ],
      processEscapes: false,
      skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
    },
    messageStyle: "normal",
    "HTML-CSS": {
      preferredFont: null
    },
    TeX: {
      extensions: ["AMSmath.js", "AMSsymbols.js","AMScd.js", "autobold.js", "begingroup.js", "mhchem.js"]
    }
  });
</script>
<script
  type="text/javascript"
  charset="utf-8"
  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"
>
</script>


  <article class="post-content">
    <div class="img_row">
	<img class="col three" src="/img/dlbio_header.png" alt="" title="Using Deep Learning Libraries to Solve Stochastic Differential Equations" />
</div>
<div class="col three caption">
</div>

<p>I am sure that most of you have heard about many of the deep learning libraries out there including <code class="language-html highlighter-rouge">TensorFlow</code>, <code class="language-html highlighter-rouge">Theano</code>, <code class="language-html highlighter-rouge">Keras</code> and <code class="language-html highlighter-rouge">PyTorch</code>. They facilitate building of layered and potentially complex neural network models for areas as diverse as automatic image captioning, speech recognition, and drug design. Undoubtedly and for better or worse, the deep learning field is riding on a wave of hype. The good thing about it is definitely that the deep learning libraries keep developing at astonishing pace. In this article, I will make use of these developments to implement two stochastic models from systems biology and solve them efficiently using the <a href="http://www.cs.cornell.edu/courses/cs5740/2017sp/lectures/04-nn-compgraph.pdf">computational graph paradigm</a> of <code class="language-html highlighter-rouge">Theano</code> and <code class="language-html highlighter-rouge">TensorFlow</code>.</p>

<h2 id="generalized-diffusion-process">Generalized diffusion process</h2>
<p><br />
In systems biology, quantitative finance, statistical physics and beyond, one often needs to solve a <em>stochastic differential equation (SDE)</em> to obtain a trustworthy model of the dynamics of studied system. This equation often takes the form of a generalized diffusion process:</p>

\[dX(t) = A\big(X(t)\big) \cdot dt + B\big(X(t)\big) \cdot W(t) \tag{1}\]

<p>where the first term on the right-hand side is <em>drift term</em> and the second term is <em>diffusion term</em>. You can think about this as pouring few drops of dyed liquid into a river stream. The color will both be carried along with the current (drift) and, to great delight of all witnessing environmental activists, gradually spread across the river width (diffusion). Diffusion is a random process and this is why the model term includes \(W(t)\), <a href="https://en.wikipedia.org/wiki/Wiener_process"><em>Wiener process</em></a> which follows normal distribution with mean \(0\) and variance \(\Delta t\) (white noise).</p>

<p>In my <a href="http://www.martinholub.com/eth/code/2018/04/15/stochsim.html">previous blog post</a>, I discussed modeling of biochemical reaction networks using the <em>chemical master equation</em> (CME). One of its drawbacks was significant computational cost. It turns out that the equation \((1)\) is a <em>Fokker-Planck equation</em>, continuous approximation of the system state dynamics (goodbye copy numbers, welcome back concentrations) and as such, it is cheaper to draw samples from.</p>

<h2 id="numerical-methods-for-sdes">Numerical methods for SDEs</h2>
<p><br />
Solution to such type of equations can be found analytically only in rare (and usually pretty boring) cases. Luckily, we can adapt some of the workhorses of numerical mathematics for solutions of ordinary differential equations (ODEs) like <a href="https://render.githubusercontent.com/view/ipynb?commit=db14f71f8b87cd9292cd15425de9bf9a512f4584&amp;enc_url=68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f6d617274696e686f6c75622f73746f636861737469632d70726f6365737365732d6b796f746f5578303039782f646231346637316638623837636439323932636431353432356465396266396135313266343538342f30312e6970796e62&amp;nwo=martinholub%2Fstochastic-processes-kyotoUx009x&amp;path=01.ipynb&amp;repository_id=129758771&amp;repository_type=Repository#Euler-Scheme"><em>Euler</em> or <em>Runge-Kutta</em></a> methods to obtain a numerical solution.</p>

<h4 id="euler-method">Euler method</h4>
<p><br />
The <strong>Euler</strong> method (also <a href="https://en.wikipedia.org/wiki/Euler%E2%80%93Maruyama_method"><strong>Euler-Maruyama</strong></a>) is straightforward extension of the deterministic variant that you have likely already heard of. Namely, given initial state \(x(0)\) it obtains consecutive states with the following rule:</p>

\[x(t+1) - x(t) = A\big(x(t)\big) \delta t + B\big(x(t)\big) \xi \delta t \tag{2}\]

<p>with \(\delta t\)</p>

\[\delta t = \frac{t_{total}}{n_{steps}}\]

<p>being size of the time interval and \(\xi \) standard normal random numbers.</p>

<p>Being as straightforward as it is, the method has <a href="http://snovit.math.umu.se/~david/Des2010/em_basel.pdf">strong convergence</a> only of the order of \(\delta t ^{\frac{1}{2}}\) (this basically says that to get twice as small error you need to make your time steps four times that small), which is poor.</p>

<h4 id="4th-order-runge-kutta-method">4th order Runge-Kutta method</h4>
<p><br />
To come up with an improvement, we have to look no further than at the familiar <a href="https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_method_(SDE)"><strong>4th order Runge-Kutta</strong></a> numerical scheme that offers strong convergence of the order of \(\delta t \). This is why I will be using this method in the following examples. The implementation here is slightly more involved, so checkout the code itself or the <a href="http://people.sc.fsu.edu/~jburkardt/m_src/stochastic_rk/rk4_ti_step.m">reference</a> if you are interested in details.</p>

<h2 id="biological-system">Biological system</h2>
<p><br />
Now that I have described the tools for solving the involved equations, let’s take a look at some examples of biological systems and model their dynamics!</p>

<h3 id="bistable-gene-circuit">Bistable gene circuit</h3>
<p><br />
Let’s consider a system composed of two genes that express transcription factors that inhibit each other (see figure 1). Careful analysis shows that such system has three equilibria, one of which being unstable. As a result, if I initialize the system close to the unstable one, it will eventually tend to one of the two stable points. At these two equilibrium points, concentration of one of the species is low whereas of the other is high. (Think about it as trying to fit a clone of Chuck Norris into a room where the original one already resides. Any room would be too small to contain awesomeness of both and thus one needs to leave for the other to thrive.)</p>

<div class="img_row">
	<img class="col three" src="/img/bistable_circuit.png" alt="" title="Bistable Gene Circuit" />
</div>
<div class="col three caption">
(a) Schematics of bistable gene circuit with two genes. (b) Phase diagram showing several trajectories converging to either one of the two stable equilibria. Along the dashed line, concentrations of both expressed transcription factors are equal, corresponding to unstable equilibrium. [1]
</div>

<p>The system is described by following pair of equations:</p>

\[\frac{x_1}{dt} = \frac{\beta_1}{1+(x_2 / K_2)^n} - \gamma x_1
\qquad
\frac{x_2}{dt} = \frac{\beta_2}{1+(x_1 / K_1)^n} - \gamma x_2
\tag{3}\]

<p>where \(\beta_i\), \(K_i\) and \(\gamma\) are some constants (which values are not really important as long as they provide for \(\dot{x}_i \approx 0\) around the the unstable equilibrium) and \(n\) is the <em>Hill coefficient</em>. I will set \(n \geq 2\) which assures <em>cooperativity</em> of binding sites (which serves the purpose of making the response of the system faster).</p>

<p>Starting from unstable initial condition \(x_1 = x_2\), I let the noise (the Wiener process term in Fokker-Planck equation) determine which stable equilibrium will the system state find.</p>

<h3 id="the-repressilator">The Repressilator</h3>
<p><br />
The “<em>repressilator</em>” is the simplest model of genetic circuit that permits sustained oscillations. It is due to the pioneering work of Elowitz and Leibler (<a href="http://www.elowitz.caltech.edu/publications/Repressilator.pdf">Nature, 2000</a>) that devised the circuit and implemented it in living bacteria. You may think of this as very simple system, but be aware that oscillatory patterns are indeed present in nature and crucial to how we as organisms function. Prominent example being <a href="https://en.wikipedia.org/wiki/Circadian_rhythm">circadian rhythm</a> that makes us function optimally on 24-hour day with sleeping time during the night (elucidation of its mechanism was honored by <a href="https://en.wikipedia.org/wiki/Nobel_Prize_in_Physiology_or_Medicine">2017 Nobel prize</a>).</p>

<p>The model itself is not anyhow more complex than the one for bistable gene circuit. In fact, the repressilator (only one of multiple setups that permit oscillations), can be obtained from the bistable genetic circuit just by adding third repressing element. This gives the following set of equations:</p>

\[\frac{x_i}{dt} = \frac{\beta_i}{1+(x_j / K_j)^n} - \gamma x_i
\qquad
i = {1,2,3},\quad j = (i + 1)~\%~3
\tag{4}\]

<div class="img_row">
	<img class="col three" src="/img/repressilator_full2.png" alt="" title="The Represillator Circuit" />
</div>
<div class="col three caption">
<i>Left</i>: Schematics of the represillator circuit where three genes express transcription factors that repress each other in turn. <i>Right</i>: An original simulation (Elowitz, 2000) showing oscillations of individual protein concentrations. Here with specific proteins cI, lacI, tetR replacing the general A, B, C notation. [1]
</div>

<h2 id="simulation-with-theano-and-tensorflow">Simulation with Theano and TensorFlow</h2>
<p><br />
As I am using a stochastic method for modelling the two systems, I will have to generate many individual realizations to obtain good representation of the system’s behavior. In order to do this efficiently, I will leverage the deep learning libraries that come with optimized implementations of <a href="https://en.wikipedia.org/wiki/Basic_Linear_Algebra_Subprograms">BLAS</a> (Basic Linear Algebra Subprograms) libraries. In particular, I can run tens of thousands of individual paths simultaneously in just one or two seconds and that on my clunky old laptop’s CPU. This is nice, especially if you need to debug your model and could use short development cycles or if your model comprises many species.</p>

<h4 id="selection-of-dl-library">Selection of DL library</h4>
<p><br />
When it comes to selection of a deep learning library, one has multiple choices. As here I am not building any complex deep learning model but rather want to use the low level functions, I have the main choice between <a href="http://deeplearning.net/software/theano/"><em>Theano</em></a> and <a href="https://www.tensorflow.org/"><em>TensorFlow</em></a>. I find the former slightly more intuitive to use and this why it would be my initial choice. However, late 2017 its core developers at LISA/MILA lab at the University of Montreal have announced that they <a href="https://groups.google.com/forum/#!msg/theano-users/7Poq8BZutbY/rNCIfvAEAwAJ">cease its further development</a>. This is why I also implemented a version that runs on <code class="language-html highlighter-rouge">Tensorflow</code>, currently the most popular deep learning framework (as per number of stars on <a href="https://github.com">github</a>).</p>

<h3 id="building-and-running-a-computational-graph">Building and running a computational graph</h3>
<p><br />
As I have already briefly mentioned in the introduction, the main point with using libraries like <code class="language-html highlighter-rouge">Theano</code> and <code class="language-html highlighter-rouge">TensorFlow</code> is that they implement different approach to computations compared to what you are probably used to. This can get pretty complicated so just keep in mind that working with <em>computational graphs</em> in these libraries boils down to two phases: <em>building</em> and <em>execution</em>.</p>

<h4 id="building">Building</h4>
<p><br />
Specifically, both libraries first express the mathematical operations symbolically, in a <strong>computational graph</strong> and optimize it for better performance and memory usage.  Computational graph (or <em>data flow graph</em>) is a directed acyclic graph (<a href="https://en.wikipedia.org/wiki/Directed_acyclic_graph">DAG</a>) in which nodes correspond to mathematical operations (“<em>Ops</em>”) and edges correspond to flow of intermediary values in between the nodes. The graph structure defines the dependencies between the operations (each node’s output depends only on the node itself and its incoming edges) and thus their order of execution <a href="http://u.cs.biu.ac.il/~yogo/nnlp.pdf">[2]</a>.</p>

<h4 id="execution">Execution</h4>
<p><br />
Once the graph has been built (which can take some time as here is where the optimization happens), computing the <em>forward pass</em> through it is straightforward and usually quite swift. It amounts to traversing the nodes and computing the output of each node given the outputs of its predecessors. If parts of the graph are independent from the rest of the structure, they can be evaluated in parallel.</p>

<p>Great, now you know the basics of working with computational graphs! Let’s go back to our examples from systems biology and build some models! Here, I will focus on the higher level implementation and results and omit the unnecessary details. Any time you feel like you need more context, you can grab the code in the <a href="https://github.com/martinholub/demos-blogs-examples/tree/master/dl-bio">github repo</a> and have some fun with it.</p>

<h3 id="bistable-gene-circuit-with-theano">Bistable gene circuit with Theano</h3>
<p><br />
First, I import the modules, including custom classes <code class="language-html highlighter-rouge">Simulator</code> and <code class="language-html highlighter-rouge">ModelParameters</code> (see full code on github), and set up some basic plotting parameters. Next, I initialize the <code class="language-html highlighter-rouge">ModelParameters</code> and <code class="language-html highlighter-rouge">Simulator</code> class objects with some parameters for the simulation</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">%</span><span class="n">load_ext</span> <span class="n">autoreload</span>
<span class="o">%</span><span class="n">autoreload</span> <span class="mi">2</span>
<span class="kn">import</span> <span class="nn">theano</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">from</span> <span class="nn">sde_theano</span> <span class="kn">import</span> <span class="n">Simulator</span><span class="p">,</span> <span class="n">ModelParameters</span>
<span class="kn">import</span> <span class="nn">matplotlib.gridspec</span> <span class="k">as</span> <span class="n">gridspec</span>

<span class="n">plt</span><span class="p">.</span><span class="n">style</span><span class="p">.</span><span class="n">use</span><span class="p">(</span><span class="s">'ggplot'</span><span class="p">)</span> <span class="c1"># use "ggplot" style for graphs
</span><span class="n">pltparams</span> <span class="o">=</span> <span class="p">{</span><span class="s">'legend.fontsize'</span><span class="p">:</span><span class="mi">14</span><span class="p">,</span><span class="s">'axes.labelsize'</span><span class="p">:</span><span class="mi">18</span><span class="p">,</span><span class="s">'axes.titlesize'</span><span class="p">:</span><span class="mi">18</span><span class="p">,</span>
             <span class="s">'xtick.labelsize'</span><span class="p">:</span><span class="mi">12</span><span class="p">,</span><span class="s">'ytick.labelsize'</span><span class="p">:</span><span class="mi">12</span><span class="p">,</span><span class="s">'figure.figsize'</span><span class="p">:(</span><span class="mf">7.5</span><span class="p">,</span><span class="mf">7.5</span><span class="p">),}</span>
<span class="n">plt</span><span class="p">.</span><span class="n">rcParams</span><span class="p">.</span><span class="n">update</span><span class="p">(</span><span class="n">pltparams</span><span class="p">)</span>

<span class="n">theano</span><span class="p">.</span><span class="n">__version__</span><span class="p">.</span><span class="n">split</span><span class="p">(</span><span class="s">"-"</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</code></pre></div></div>

<div class="language-html highlighter-rouge"><div class="highlight"><pre class="highlight"><code>'0.9.0.dev'
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">total_time</span> <span class="o">=</span> <span class="mi">9</span>
<span class="n">num_samples</span> <span class="o">=</span> <span class="mf">5e4</span>
<span class="n">dt0</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="n">model</span> <span class="o">=</span> <span class="s">"switch"</span>
<span class="n">mp</span> <span class="o">=</span> <span class="n">ModelParameters</span><span class="p">(</span><span class="n">num_samples</span><span class="p">,</span> <span class="n">model_type</span> <span class="o">=</span> <span class="n">model</span><span class="p">,</span> <span class="n">dt0</span> <span class="o">=</span> <span class="n">dt0</span><span class="p">,</span>
                     <span class="n">t_end</span> <span class="o">=</span> <span class="n">total_time</span><span class="p">,</span> <span class="n">nspecies</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">sde_solver</span> <span class="o">=</span> <span class="n">Simulator</span><span class="p">(</span><span class="n">seed</span> <span class="o">=</span> <span class="mi">56</span><span class="p">,</span> <span class="n">rv_shape</span> <span class="o">=</span> <span class="n">mp</span><span class="p">.</span><span class="n">x0</span><span class="p">.</span><span class="n">shape</span><span class="p">)</span>
</code></pre></div></div>

<p>As introduced above, I first build a computational graph of the mathematical operations using <a href="http://deeplearning.net/software/theano/library/scan.html"><code class="language-html highlighter-rouge">theano.scan</code></a> that is used for defining looping-like operations in the graph. Next, I compile it into a Python callable <code class="language-html highlighter-rouge">sim</code> with <a href="http://deeplearning.net/software/theano/library/compile/function.html"><code class="language-html highlighter-rouge">theano.function</code></a> such that I can call it as any other function.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Define loop: symbolic description of the result and how to obtain it
# x_out ... 3D Tensor, updates ... dict ,
</span><span class="p">(</span><span class="n">x_out</span><span class="p">,</span><span class="n">updates</span><span class="p">)</span><span class="o">=</span><span class="n">theano</span><span class="p">.</span><span class="n">scan</span><span class="p">(</span><span class="n">fn</span> <span class="o">=</span> <span class="n">sde_solver</span><span class="p">.</span><span class="n">rk4_sde</span><span class="p">,</span>   <span class="c1">#stepping function
</span>                            <span class="n">outputs_info</span> <span class="o">=</span> <span class="p">[</span><span class="n">mp</span><span class="p">.</span><span class="n">x</span><span class="p">],</span>     <span class="c1">#output shape
</span>                            <span class="c1"># set taps to [-1], dont specify IC
</span>                            <span class="n">non_sequences</span> <span class="o">=</span> <span class="n">mp</span><span class="p">.</span><span class="n">params</span><span class="p">,</span> <span class="c1">#fixed parameters
</span>                            <span class="n">n_steps</span> <span class="o">=</span> <span class="n">mp</span><span class="p">.</span><span class="n">nsteps</span><span class="p">)</span>       <span class="c1"># number of iterations
</span>
<span class="c1"># Compile symbolic graph representation into callable function
</span><span class="n">sim</span><span class="o">=</span><span class="n">theano</span><span class="p">.</span><span class="n">function</span><span class="p">(</span><span class="n">inputs</span><span class="o">=</span><span class="n">mp</span><span class="p">.</span><span class="n">params</span><span class="p">,</span>     <span class="c1"># args to function
</span>                    <span class="n">outputs</span><span class="o">=</span><span class="n">x_out</span><span class="p">,</span>        <span class="c1"># symbolic outputs
</span>                    <span class="n">givens</span><span class="o">=</span><span class="p">{</span><span class="n">mp</span><span class="p">.</span><span class="n">x</span><span class="p">:</span><span class="n">mp</span><span class="p">.</span><span class="n">x0</span><span class="p">},</span>  <span class="c1"># set initial value
</span>                    <span class="n">updates</span><span class="o">=</span><span class="n">updates</span><span class="p">,</span>      <span class="c1"># how to obtain the outputs
</span>                    <span class="n">allow_input_downcast</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                    <span class="n">profile</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>         <span class="c1"># get time spent in ops
</span></code></pre></div></div>

<p>Subsequently, I can execute the compiled graph as usual. I also use <a href="http://www.graphviz.org/"><code class="language-html highlighter-rouge">graphviz</code></a> to obtain nice visual rendering of the nodes and edges in the graph and even the results of profiling the code execution.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="s">"running sim..."</span><span class="p">)</span>
<span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="p">.</span><span class="n">clock</span><span class="p">()</span>
<span class="n">x_out</span> <span class="o">=</span> <span class="n">sim</span><span class="p">(</span><span class="o">*</span><span class="n">mp</span><span class="p">.</span><span class="n">params0</span><span class="p">)</span>
<span class="n">diff</span> <span class="o">=</span> <span class="p">(</span><span class="n">time</span><span class="p">.</span><span class="n">clock</span><span class="p">()</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s">"done in"</span><span class="p">,</span> <span class="n">diff</span><span class="p">,</span> <span class="s">"s at "</span><span class="p">,</span> <span class="n">diff</span><span class="o">/</span><span class="n">num_samples</span><span class="p">,</span> <span class="s">"s per path"</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-html highlighter-rouge"><div class="highlight"><pre class="highlight"><code>running sim...
done in 3.0696229474693153 s at  6.13924589493863e-05 s per path
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">from</span> <span class="nn">theano.printing</span> <span class="kn">import</span> <span class="n">pydotprint</span>
<span class="kn">from</span> <span class="nn">theano.d3viz</span> <span class="kn">import</span> <span class="n">d3viz</span>

<span class="n">pydotprint</span><span class="p">(</span><span class="n">sim</span><span class="p">,</span> <span class="s">"pics/switch_graph.png"</span><span class="p">,</span> <span class="n">var_with_name_simple</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">d3viz</span><span class="p">(</span><span class="n">sim</span><span class="p">,</span> <span class="s">"pics/switch_graph.html"</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-html highlighter-rouge"><div class="highlight"><pre class="highlight"><code>The output file is available at pics/switch_graph.png
</code></pre></div></div>

<p>Here is the resulting visualization and here you can check the interactive <a href="/img/switch_graph.html" target="_blank"><code class="language-html highlighter-rouge">html</code></a> output.</p>

<div class="img_row">
    <a href="/img/switch_graph.png" target="_blank">
	<img class="col three" src="/img/switch_graph.png" alt="Graph Switch" title="Graph Switch" /></a>
</div>
<div class="col three caption">
Image of the computational graph with nodes representing mathematical operations or variables and edges representing flow of data ("tensors").
</div>

<p>Next, let’s plot the results:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">## for model SWITCH , nspecies = 2
</span>
<span class="n">downsample_factor_t</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mf">0.1</span><span class="o">/</span><span class="n">dt0</span><span class="p">)</span> <span class="c1"># show 10 points per time unit, here no downs.
</span><span class="n">downsample_factor_p</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">num_samples</span><span class="o">//</span><span class="mi">50</span><span class="p">)</span> <span class="c1">## show every 50th path
</span><span class="n">t</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">total_time</span><span class="p">,</span> <span class="n">mp</span><span class="p">.</span><span class="n">nsteps</span><span class="o">//</span><span class="n">downsample_factor_t</span><span class="p">)</span> <span class="c1"># time axis
</span>
<span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="p">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="n">plt</span><span class="p">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="c1"># plot just first of the two species
</span><span class="n">plt</span><span class="p">.</span><span class="n">plot</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">x_out</span><span class="p">[::</span><span class="n">downsample_factor_t</span><span class="p">,</span> <span class="p">::</span><span class="n">downsample_factor_p</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>

<span class="n">plt</span><span class="p">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.2</span><span class="p">,</span> <span class="mi">50</span><span class="p">)</span>
<span class="n">plt</span><span class="p">.</span><span class="n">hist</span><span class="p">(</span><span class="n">x_out</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="n">dt0</span><span class="p">),</span> <span class="p">:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">bins</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
            <span class="n">normed</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">histtype</span><span class="o">=</span><span class="s">'bar'</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="p">[</span><span class="s">'t = 1'</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">hist</span><span class="p">(</span><span class="n">x_out</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="mi">2</span><span class="o">/</span><span class="n">dt0</span><span class="p">),</span> <span class="p">:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">bins</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
            <span class="n">normed</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">histtype</span><span class="o">=</span><span class="s">'bar'</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="p">[</span><span class="s">'t = 2'</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">hist</span><span class="p">(</span><span class="n">x_out</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="p">:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">bins</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
            <span class="n">normed</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">histtype</span><span class="o">=</span><span class="s">'bar'</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="p">[</span><span class="s">'t = 9'</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></div>

<div class="img_row" style="width: 80%;">
	<img class="col three" src="/img/switch_out.png" alt="" title="Switch Plots" />
</div>
<div class="col three caption">
</div>

<p>In the plot, you can clearly see that the system follows the announced behavior. The concentration of any of the two species (just one shown) starts in the unstable equilibrium  and, based on the stochastic input, it evolves either to the high or the low stable equilibrium state. In the lower panel, you can see how the distribution of the concentration values changes in time. The example may seem trivial, but this is <a href="https://doi.org/10.1016/j.cell.2011.01.030">indeed one of the mechanisms of cellular differentiation</a> permitting nongenetic cellular diversity. All things being equal, whether cell takes up fate A or B can be matter of pure chance (given our current level of understanding of determinism in biological systems :) )</p>

<h3 id="the-repressilator-with-tensorflow">The Repressilator with TensorFlow</h3>
<p><br />
For the repressilator model, I will make use of <code class="language-html highlighter-rouge">TensorFlow</code>’s implementation of <a href="https://www.tensorflow.org/api_docs/python/tf/scan"><code class="language-html highlighter-rouge">scan</code></a> function. Similar to the <code class="language-html highlighter-rouge">Theano</code>’s version, <code class="language-html highlighter-rouge">scan</code> allows to define loops in the computational graph paradigm.
First, we again import the required modules. Note that here I have replaced the custom <code class="language-html highlighter-rouge">sde_theano</code> module with <code class="language-html highlighter-rouge">sde_tflow</code> (full code on github). I then initialize instances of the custom classes <code class="language-html highlighter-rouge">ModelParameters</code> and <code class="language-html highlighter-rouge">Simulator</code> with some parameters for the simulation.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">%</span><span class="n">load_ext</span> <span class="n">autoreload</span>
<span class="o">%</span><span class="n">autoreload</span> <span class="mi">2</span>
<span class="kn">import</span> <span class="nn">tensorflow</span> <span class="k">as</span> <span class="n">tf</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">from</span> <span class="nn">sde_tflow</span> <span class="kn">import</span> <span class="n">Simulator</span><span class="p">,</span> <span class="n">ModelParameters</span>
<span class="kn">import</span> <span class="nn">matplotlib.gridspec</span> <span class="k">as</span> <span class="n">gridspec</span>
<span class="kn">import</span> <span class="nn">time</span>

<span class="n">plt</span><span class="p">.</span><span class="n">style</span><span class="p">.</span><span class="n">use</span><span class="p">(</span><span class="s">'ggplot'</span><span class="p">)</span> <span class="c1"># use "ggplot" style for graphs
</span><span class="n">pltparams</span> <span class="o">=</span> <span class="p">{</span><span class="s">'legend.fontsize'</span><span class="p">:</span><span class="mi">14</span><span class="p">,</span><span class="s">'axes.labelsize'</span><span class="p">:</span><span class="mi">18</span><span class="p">,</span><span class="s">'axes.titlesize'</span><span class="p">:</span><span class="mi">18</span><span class="p">,</span>
             <span class="s">'xtick.labelsize'</span><span class="p">:</span><span class="mi">12</span><span class="p">,</span><span class="s">'ytick.labelsize'</span><span class="p">:</span><span class="mi">12</span><span class="p">,</span><span class="s">'figure.figsize'</span><span class="p">:(</span><span class="mf">7.5</span><span class="p">,</span><span class="mf">7.5</span><span class="p">),}</span>
<span class="n">plt</span><span class="p">.</span><span class="n">rcParams</span><span class="p">.</span><span class="n">update</span><span class="p">(</span><span class="n">pltparams</span><span class="p">)</span>

<span class="n">tf</span><span class="p">.</span><span class="n">__version__</span>
</code></pre></div></div>

<div class="language-html highlighter-rouge"><div class="highlight"><pre class="highlight"><code>'1.5.1'
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">total_time</span> <span class="o">=</span> <span class="mi">9</span>
<span class="n">num_samples</span> <span class="o">=</span> <span class="mf">2.5e4</span>
<span class="n">dt0</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="n">model</span> <span class="o">=</span> <span class="s">"rep"</span>
<span class="n">mp</span> <span class="o">=</span> <span class="n">ModelParameters</span><span class="p">(</span> <span class="n">num_samples</span><span class="p">,</span> <span class="n">model_type</span> <span class="o">=</span> <span class="n">model</span><span class="p">,</span>
                      <span class="n">dt0</span> <span class="o">=</span> <span class="n">dt0</span><span class="p">,</span> <span class="n">t_end</span> <span class="o">=</span> <span class="n">total_time</span><span class="p">)</span>
<span class="n">sde_solver</span> <span class="o">=</span> <span class="n">Simulator</span><span class="p">(</span><span class="n">mp</span><span class="p">,</span> <span class="n">seed</span> <span class="o">=</span> <span class="mi">56</span><span class="p">)</span>
</code></pre></div></div>

<p>Next, I build the computational graph and let <code class="language-html highlighter-rouge">TensorFlow</code> optimize its structure. As I am using similar functionality as <code class="language-html highlighter-rouge">Theano</code> the call does not differ much. I supply the function to evaluate at each execution, initial state and elements to loop over.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Define loop: symbolic description of the result and how to obtain it
</span><span class="n">x_scan</span> <span class="o">=</span> <span class="n">tf</span><span class="p">.</span><span class="n">scan</span><span class="p">(</span><span class="n">fn</span> <span class="o">=</span> <span class="n">sde_solver</span><span class="p">.</span><span class="n">rk4_sde</span><span class="p">,</span>        <span class="c1"># stepping function
</span>                    <span class="n">elems</span> <span class="o">=</span> <span class="n">sde_solver</span><span class="p">.</span><span class="n">rv_n</span><span class="p">,</span>     <span class="c1"># inputs for all steps
</span>                    <span class="n">initializer</span> <span class="o">=</span> <span class="n">sde_solver</span><span class="p">.</span><span class="n">x0</span><span class="p">)</span> <span class="c1"># initial condition
</span></code></pre></div></div>

<p>With the graph built, I proceed to its execution (the forward pass through it). In <code class="language-html highlighter-rouge">TensorFlow</code>, you run a graph in a <a href="https://www.tensorflow.org/programmers_guide/graphs#executing_a_graph_in_a_wzxhzdk132wzxhzdk133tfsessionwzxhzdk134wzxhzdk135"><code class="language-html highlighter-rouge">tf.Session()</code></a> that encapsulates the environment the operations and tensors live in. Since a <code class="language-html highlighter-rouge">tf.Session</code> owns a physical  resources, it is recommendable to use it as a context manager (in a <code class="language-html highlighter-rouge">with</code> block) that automatically closes the session at the exit of the block.</p>

<p>Note that I also write the graph that is run in the session to a file in <code class="language-html highlighter-rouge">log</code> folder. This graph can then be interactively visualized with <a href="https://www.tensorflow.org/programmers_guide/summaries_and_tensorboard"><code class="language-html highlighter-rouge">TensorBoard</code></a>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="s">"running sim..."</span><span class="p">)</span>
<span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="p">.</span><span class="n">clock</span><span class="p">()</span>

<span class="k">with</span> <span class="n">tf</span><span class="p">.</span><span class="n">Session</span><span class="p">()</span> <span class="k">as</span> <span class="n">sess</span><span class="p">:</span> <span class="c1"># open a Session
</span>    <span class="c1"># Write the graph to file
</span>    <span class="n">writer</span> <span class="o">=</span> <span class="n">tf</span><span class="p">.</span><span class="n">summary</span><span class="p">.</span><span class="n">FileWriter</span><span class="p">(</span><span class="s">"log/"</span><span class="p">,</span> <span class="n">sess</span><span class="p">.</span><span class="n">graph</span><span class="p">)</span>
    <span class="n">x_out</span> <span class="o">=</span> <span class="n">sess</span><span class="p">.</span><span class="n">run</span><span class="p">(</span><span class="n">x_scan</span><span class="p">)</span> <span class="c1"># execute the graph
</span>    <span class="n">writer</span><span class="p">.</span><span class="n">close</span><span class="p">()</span> <span class="c1"># close the FileWriter
</span>
<span class="n">diff</span> <span class="o">=</span> <span class="p">(</span><span class="n">time</span><span class="p">.</span><span class="n">clock</span><span class="p">()</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s">"done in"</span><span class="p">,</span> <span class="n">diff</span><span class="p">,</span> <span class="s">"s at "</span><span class="p">,</span> <span class="n">diff</span><span class="o">/</span><span class="n">num_samples</span><span class="p">,</span> <span class="s">"s per path"</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-html highlighter-rouge"><div class="highlight"><pre class="highlight"><code>running sim...
done in 3.599450070369931 s at  0.00014397800281479723 s per path
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">## for model REPRESILLATOR
</span>
<span class="n">downsample_factor_t</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mf">0.1</span><span class="o">/</span><span class="n">dt0</span><span class="p">)</span> <span class="c1">#always show 10 points per time unit
</span><span class="n">downsample_factor_p</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">num_samples</span><span class="o">//</span><span class="mi">50</span><span class="p">)</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">total_time</span><span class="p">,</span> <span class="n">mp</span><span class="p">.</span><span class="n">nsteps</span><span class="o">//</span><span class="n">downsample_factor_t</span><span class="p">)</span>

<span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="p">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">4</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>

<span class="n">plt</span><span class="p">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">plot</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">x_out</span><span class="p">[::</span><span class="n">downsample_factor_t</span><span class="p">,</span> <span class="p">::</span><span class="n">downsample_factor_p</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">plot</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">x_out</span><span class="p">[::</span><span class="n">downsample_factor_t</span><span class="p">,</span> <span class="p">::</span><span class="n">downsample_factor_p</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">plot</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">x_out</span><span class="p">[::</span><span class="n">downsample_factor_t</span><span class="p">,</span> <span class="p">::</span><span class="n">downsample_factor_p</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>

<span class="n">plt</span><span class="p">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">hist</span><span class="p">(</span><span class="n">x_out</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,:,</span><span class="mi">0</span><span class="p">],</span> <span class="mi">30</span><span class="p">,</span>
            <span class="n">normed</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">histtype</span><span class="o">=</span><span class="s">'bar'</span><span class="p">)</span>
<span class="n">plt</span><span class="p">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">hist</span><span class="p">(</span><span class="n">x_out</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,:,</span><span class="mi">1</span><span class="p">],</span> <span class="mi">30</span><span class="p">,</span>
            <span class="n">normed</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">histtype</span><span class="o">=</span><span class="s">'bar'</span><span class="p">)</span>
<span class="n">plt</span><span class="p">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">hist</span><span class="p">(</span><span class="n">x_out</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,:,</span><span class="mi">2</span><span class="p">],</span> <span class="mi">30</span><span class="p">,</span>
            <span class="n">normed</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">histtype</span><span class="o">=</span><span class="s">'bar'</span><span class="p">)</span>
<span class="n">plt</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></div>

<div class="img_row" style="width: 80%;">
	<img class="col three" src="/img/repressilator_out.png" alt="" title="Repressilator Plots" />
</div>
<div class="col three caption">
</div>

<p>Here are the results. You can nicely see the oscillatory pattern that however slowly degrades in time (this makes sense as the noise makes the paths more different). The paths for individual species are phase shifted such that each of the three reaches its maximum at different time. You can even save the histograms at couple of time steps and create an animation like this:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">## Save plots for animation
</span><span class="n">plt</span><span class="p">.</span><span class="n">clf</span><span class="p">()</span>
<span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">linspace</span><span class="p">(</span><span class="mf">0.1</span> <span class="p">,</span> <span class="mf">1.1</span><span class="p">,</span> <span class="mi">50</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">np</span><span class="p">.</span><span class="n">arange</span><span class="p">(</span><span class="n">x_out</span><span class="p">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="n">plt</span><span class="p">.</span><span class="n">hist</span><span class="p">(</span><span class="n">x_out</span><span class="p">[</span><span class="n">i</span><span class="p">,:,</span><span class="mi">1</span><span class="p">],</span> <span class="n">bins</span><span class="p">,</span>
            <span class="n">normed</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">histtype</span><span class="o">=</span><span class="s">'bar'</span><span class="p">)</span>
    <span class="n">plt</span><span class="p">.</span><span class="n">xlim</span><span class="p">([</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">1.1</span><span class="p">])</span>
    <span class="n">plt</span><span class="p">.</span><span class="n">ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">])</span>
    <span class="n">plt</span><span class="p">.</span><span class="n">savefig</span><span class="p">(</span><span class="s">"pics/tf/rep"</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s">".png"</span><span class="p">)</span>
    <span class="n">plt</span><span class="p">.</span><span class="n">clf</span><span class="p">()</span>
</code></pre></div></div>

<div class="img_row" style="width: 80%;">
	<img class="col three" src="/img/repressilator_anim.gif" alt="Repressilator Animation" title="Repressilator Animation" />
</div>
<div class="col three caption">
Animation showing temporal evolution of concentrations of protein B in the repressilator model.
</div>

<p>And here is a screenshot from the <code class="language-html highlighter-rouge">TensorBoard</code> application. Unfortunately it is quite crowded and thus not well readable (it is much better inside <code class="language-html highlighter-rouge">TensorBoard</code> as the graph is interactive).</p>

<div class="img_row">
    <a href="/img/tensorboard_repressilator.png" target="_blank">
	<img class="col three" src="/img/tensorboard_repressilator.png" alt="" title="Tensorboard Represillator" /></a>
</div>
<div class="col three caption">
Image of the computational graph with nodes representing mathematical operations or variables and edges representing flow of data ("tensors") as rendered by TensorBoard.
</div>

<h2 id="conclusion">Conclusion</h2>
<p><br />
In this post, I have shown you how to efficiently solve stochastic differential equations using computational graph approach implemented in deep learning libraries like <code class="language-html highlighter-rouge">Theano</code> or <code class="language-html highlighter-rouge">TensorFlow</code>. You can now develop these examples further and build a stochastic model of your own or even implement a neural network model in <a href="https://caffe2.ai/docs/applications-of-deep-learning.html">virtually any domain of your interest</a>. If you will do so, let me know, I am interested to hear about your experience.</p>

<hr />

<p><br /></p>
<h3 id="references">References</h3>
<p><br /></p>

<p>[1]: Del Vecchio, D., Murray, R. M.; Biomolecular Feedback Systems<br />
  [2]: Goldber, Y.; A Primer on Neural Network Models for Natural Language Processing</p>

  </article>

  
<!---
Call Fontawesome in the head section or in the location where you place the share bar
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
--->
<hr>
<br/>
<!---<center><p><script async src="https://eocampaign1.com/form/cfe45fdc-109a-11ef-a9e2-bf0899395051.js" data-form="cfe45fdc-109a-11ef-a9e2-bf0899395051"></script></p></center>--->
<div class="share-box center" style="width: 60%; margin: 0px auto;">
<h2 class="center">Share this:</h2>
<br/>
<a class="l" href="https://www.linkedin.com/shareArticle?mini=true&url=https://www.martinholub.com/2018/05/01/TheanoTensorflowSDE.html&title=Using Deep Learning Libraries to Solve Stochastic Differential Equations&summary=&source=martinholub" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fa-brands fa-linkedin"></i><span> linkedin</span></a>

<a class="t" href="https://twitter.com/intent/tweet?text=Using Deep Learning Libraries to Solve Stochastic Differential Equations&url=https://www.martinholub.com/2018/05/01/TheanoTensorflowSDE.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fa-brands fa-square-x-twitter"></i><span> x</span></a>

<a class="t" href="https://bsky.app/intent/post?text=Using Deep Learning Libraries to Solve Stochastic Differential Equations&url=https://www.martinholub.com/2018/05/01/TheanoTensorflowSDE.html" onclick="window.open(this.href, 'mywin', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;">
  <i class="fa-brands fa-square-bluesky"></i><span> bluesky</span>
</a>
<!---
<a class="g" href="https://plus.google.com/share?url=https://www.martinholub.com/2018/05/01/TheanoTensorflowSDE.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" ><i class="fa fa-google-plus fa"></i><span> google</span></a>
--->
<a class="r" href="http://www.reddit.com/submit?url=https://www.martinholub.com/2018/05/01/TheanoTensorflowSDE.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=900,height=500,toolbar=1,resizable=0'); return false;"><i class="fa-brands fa-square-reddit"></i><span> reddit</span></a>

<a class="f" href="https://www.facebook.com/sharer/sharer.php?u=https://www.martinholub.com/2018/05/01/TheanoTensorflowSDE.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" ><i class="fa-brands fa-square-facebook"></i><span> facebook</span></a>

<a class="e" href="mailto:?subject=Using Deep Learning Libraries to Solve Stochastic Differential Equations&amp;body=Check out this site https://www.martinholub.com/2018/05/01/TheanoTensorflowSDE.html"><i class="fa fa-envelope fa"></i><span> email</span></a>
</div>
<br/>
<hr>


  

<div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'martinholub'; // required: replace example with your forum shortname
        /*var disqus_developer = 1; // Comment out when the site is live */
        var disqus_identifier = "/2018/05/01/TheanoTensorflowSDE.html";

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>



</div>

      </div>
    </div>

<!--
    <span class="contacticon center">
  <a href="https://www.instagram.com/scimartin/" target="_blank"><i class="fa-brands fa-square-instagram"></i></a>
  <a href="https://www.linkedin.com/in/scimartin" target="_blank"><i class="fa-brands fa-linkedin"></i></a>
  <a href="https://bsky.app/profile/martinholub.com" target="_blank"><i class="fa-brands fa-square-bluesky"></i></a>
  <br>
  <a href="mailto:scimartin=proton+me"><i class="fa fa-envelope-square"></i></a>
  <a href="https://orcid.org/0000-0002-8365-0927" target="_blank"><i class="ai ai-orcid"></i></a>
  <a href="https://github.com/martinholub" target="_blank"><i class="fa-brands fa-square-github"></i></a>
</span>


<div class="col three caption">
	&#169; Martin Holub, 2025
</div>
-->
  <script id="dsq-count-scr" src="//martinholub.disqus.com/count.js" async></script>
  <script src="/css/lightbox.js"></script>

  </body>
    <span class="contacticon center">
  <a href="https://www.instagram.com/scimartin/" target="_blank"><i class="fa-brands fa-square-instagram"></i></a>
  <a href="https://www.linkedin.com/in/scimartin" target="_blank"><i class="fa-brands fa-linkedin"></i></a>
  <a href="https://bsky.app/profile/martinholub.com" target="_blank"><i class="fa-brands fa-square-bluesky"></i></a>
  <br>
  <a href="mailto:scimartin=proton+me"><i class="fa fa-envelope-square"></i></a>
  <a href="https://orcid.org/0000-0002-8365-0927" target="_blank"><i class="ai ai-orcid"></i></a>
  <a href="https://github.com/martinholub" target="_blank"><i class="fa-brands fa-square-github"></i></a>
</span>


<div class="col three caption">
	&#169; Martin Holub, 2025
</div>

</html>
